---
title: "GTEx sQTL enrichment (V4, different pseudocount)"
date: 2024-8-15
categories: ['qtl']
code-fold: true
execute:
  include: true
  eval: true
  cache: false
  echo: false
---


::: {.callout-note}
Changed pseducount count method in phenotype prep script

-  from: $$\frac{num + 0.5 }{ denom + 0.5}$$
-  to  : $$\frac{num + addon }{ denom + addon}$$, where `addon = denom * .01 if denom < 100 else 0.01`


- liver/kidney plots used spearman correlation and new pseudo count
- other tissues used pearson correlation and close to old pseudo count
:::



The underlying data for both the qqplot and the scatter plot are the from this procedure:

1. Map sQTLs using permutation pass with QTLtools against leafcutter2 output. Each phenotype is an intron.
2. Permpass result report a top SNP for each intron. Use the collection of these top SNPs to subset genotype VCF file. 
3. Run nominal pass for eQTL with expression data and the subsetted genotype VCF file. This effectively run regression analysis of each top sQTL SNP against gene expression.

Then for the qqplot, I plotted the pvalue of expression regression for each productive vs. unproductive sQTLs top SNP.

For the scatter plot, each top SNP is the top SNP of a productive or unprodutive sQTL. X axis is SNP's regression slope for gene expression, and Y axis is the regression slope of splicing.

Note each sQTL is either a productive or unproductive sQTLs, and represent a intron cluster. A productive sQTL is a cluster that only have productive introns. The strongest intron is chosen to represent the cluster (instead of aggregated of all the introns). 

An unproductive sQTL is a cluster that has both productive intron and unproductive introns with permutation pass q value < 0.1. The best intron (largest top SNP slope) is chosen to represent the cluster as unproductive sQTL. Note, comparing to previous version, here for u-sQTLs (ctype = PR,UP) select the best unproductive intron to represent the cluster.

::: {.callout-note}
This update removed the following categories of SNPs from the GTEx VEP annotation file:

  - enhancer_d
  - promoter_d
  - promoter_flanking_region_d

qqplot Y axis is flipped if sQTL slope and eQTL slope are in opposite direction.
:::



```{r}
#| include: false
suppressMessages(library(data.table))
suppressMessages(library(tidyverse))
library(glue)

fdr = 0.1
```


```{r}
base_path= "../smk-plots/gtex-sqtl-enrichment-v4-test"

```

```{r}
#| include: false
#| echo: false
qqs <- list.files(base_path, pattern = "qqplot*", full.names = TRUE)
scatter <- list.files(base_path, pattern = "scatter", full.names = TRUE)
head(qqs)
head(scatter)
```


```{r}
#| include: false
#| echo: false
qqs.old <- str_replace(qqs, "v4-test", "v4")
scatter.old <- str_replace(scatter, "v4-test", "v4")
```

# Enrichment of sQTL in eQTL pvalues


## With new pseudo count

```{r}
#| column: screen
#| label: "fig-qqplot-eQTL-sQTL-diff-direction"
#| layout-ncol: 4
#| out-width: 100%
#| fig-cap: qqplot of eQTL pvalues of sQTL top SNPs (new pseudocount)

knitr::include_graphics(qqs)
```



## with old pseudo

```{r}
#| column: screen
#| label: "fig-qqplot-eQTL-sQTL-diff-direction-old"
#| layout-ncol: 4
#| out-width: 100%

knitr::include_graphics(qqs.old)
```

# Comparison of sQTL (top SNP beta) and eQTL effect size (nominal beta of same top sQTL SNP)

::: {.callout-note}

- cor.test used pearson correlation

:::


## With new pseudo count


```{r}
#| column: screen
#| label: fig-scatterplots
#| fig-cap: comparison of sQTL and eQTL effect size (new pseudocount)
#| layout-ncol: 4
#| out-width: 100%

knitr::include_graphics(scatter)
```

## With old pseudo count

```{r}
#| column: screen
#| label: fig-scatterplots-old
#| fig-cap: comparison of sQTL and eQTL effect size (old)
#| layout-ncol: 4
#| out-width: 100%

knitr::include_graphics(scatter.old)
```


# Compare number of qtls

```{r}
tissues = qqs %>% basename %>% str_remove("\\-qqplot\\.png")
names(tissues) = tissues
```


## with new pseudo count

```{r}
paths2 = map(
  tissues,
  ~glue("../code/results/qtl/noisy/GTEx/{.x}/test_pseudocount/cis_100000/perm") %>% 
    dir(pattern = ".*addQval.txt.gz$", full.names = TRUE) %>% 
    naturalsort::naturalsort(.)
)
```
  

```{r}
#| cache: true

sqtl.perms2 = map(paths2, ~map_dfr(.x, fread))
sqtl.perms2 = map(
  sqtl.perms2,
  ~.x[str_detect(phenotype_id, "PR|UP|NE") & q < fdr] %>% 
    .[, .(phenotype_id, itype = str_extract(phenotype_id, "PR|UP|NE"))]
)

map(
  sqtl.perms2,
    ~.x[, .(N = uniqueN(phenotype_id)), by = itype] %>% 
    .[, .(N, P = round(N/sum(N), 3), itype)] %>% 
    .[order(-N)]
)


```


## with old pseudo count

```{r}
paths = map(
  tissues,
  ~glue("../code/results/qtl/noisy/GTEx/{.x}/separateNoise/cis_100000/perm") %>% 
    dir(pattern = ".*addQval.txt.gz$", full.names = TRUE) %>% 
    naturalsort::naturalsort(.)
)
```
  

```{r}
#| cache: true

sqtl.perms = map(paths, ~map_dfr(.x, fread))
sqtl.perms = map(
  sqtl.perms,
  ~.x[str_detect(phenotype_id, "PR|UP|NE") & q < fdr] %>% 
    .[, .(phenotype_id, itype = str_extract(phenotype_id, "PR|UP|NE"))]
)

map(
  sqtl.perms,
    ~.x[, .(N = uniqueN(phenotype_id)), by = itype] %>% 
    .[, .(N, P = round(N/sum(N), 3), itype)] %>% 
    .[order(-N)]
)
```


## intersection of sQTLs between new and old pseudo count


```{r}
#| cache: true
intersected_ids = map2(
  sqtl.perms2,
  sqtl.perms,
  ~list(
    intersect = intersect(.x$phenotype_id, .y$phenotype_id),
    inNewOnly = setdiff(.x$phenotype_id, .y$phenotype_id),
    inOldOnly = setdiff(.y$phenotype_id, .x$phenotype_id)
  )
)

map(
  intersected_ids,
  ~map(.x, ~length(.x)) %>% as.data.table
)
```



