---
title: "sQTL procedure"
date: 2024-7-16
format: html
categories: ['sqtl', 'method']
code-fold: false
execute:
  include: true
  eval: true
  cache: false
  echo: true
---



# `QTLtools cis --perm` data preparation


1. bed: phenotype bed file
2. cov: covariate file
3. vcf: genotype vcf file


## phenotype bed file

Used a python [script](https://github.com/daiuc/SpliFi/blob/d433ad3784df09b659b8d54e8b8e2bf5f7aa280b/code/workflow/scripts/preparePheno.py) to prepare phenotype for QTLtools.


- introns with missing PSI ratios from over 40% of samples are removed.
- introns with less than 0.005 std are also removed.
- PSI values from leafcutter output are [added with pseudo count](https://github.com/daiuc/SpliFi/blob/d433ad3784df09b659b8d54e8b8e2bf5f7aa280b/code/workflow/scripts/preparePheno.py#L166C17-L166C68) of 0.5.

```{python}
#| eval: false
count = (float(num) + 0.5) / ((float(denom)) + 0.5)
```


- PSI values are quantile normalized

```{python}
#| eval: false
def qqnorm(x):
    n = len(x)
    a = 3.0 / 8.0 if n <= 10 else 0.5
    return norm.ppf((rankdata(x) - a) / (n + 1.0 - 2.0 * a))

# qqnorms on the columns
matrix = np.array(valRows)
for i in range(len(matrix[0, :])):
    matrix[:, i] = qqnorm(matrix[:, i])

```


# covariate file

A combination of:

- phenotype PCs (top 11)
- genotype pcs (top 5)





















































