{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "import glob\n",
    "from collections import OrderedDict\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.ticker as ticker\n",
    "from matplotlib.colors import hsv_to_rgb\n",
    "\n",
    "import qtl.annotation\n",
    "import qtl.plot\n",
    "import qtl.torus\n",
    "import style\n",
    "\n",
    "annot = qtl.annotation.Annotation('../data/gencode.v26.GRCh38.genes.gtf', verbose=False)\n",
    "sample_df = pd.read_csv('../code/workflow/submodules/gtex-v8/data/GTEx_Analysis_v8_RNAseq_samples.txt', sep='\\t')\n",
    "sample_s = sample_df[sample_df['has_genotype']].groupby('tissue_id').apply(lambda x: x['sample_id'].tolist())\n",
    "sample_s = sample_s[sample_s.apply(len)>=70]\n",
    "size_s = sample_s.apply(len)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/project/yangili1/cdai/splice-pub/analysis\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# VEP annotation for top eQTL/sQTL\n",
    "vep_df = pd.read_csv('../data/WGS_Feature_overlap_collapsed_VEP_short_4torus.MAF01.txt.gz', sep='\\t', index_col=0)\n",
    "vep_df = vep_df.astype(np.int8)\n",
    "vep_df.columns = [qtl.torus.torus_dict[i[:-2].upper()] for i in vep_df.columns]\n",
    "\n",
    "# variant IDs for each category\n",
    "variant_sets = {}\n",
    "for p in vep_df:\n",
    "    variant_sets[p] = set(vep_df.loc[vep_df[p]==1].index)\n",
    "variant_set_s = pd.Series(variant_sets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['Enhancer', 'Promoter', 'Open chromatin region', 'Promoter-flanking region', 'CTCF binding site', 'TF binding site', \"3' UTR variant\", \"5' UTR variant\", 'Frameshift variant', 'Intron variant', 'Missense variant', 'Non-coding transcript exon variant', 'Splice acceptor variant', 'Splice donor variant', 'Splice region variant', 'Stop gained', 'Synonymous variant'])"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "82219"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "29762"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "['chr6_125781340_C_G_b38',\n",
       " 'chr8_115623645_T_TA_b38',\n",
       " 'chr10_92691973_C_A_b38',\n",
       " 'chrX_49164178_G_T_b38',\n",
       " 'chr8_33513467_A_G_b38']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check length of variant_sets\n",
    "variant_sets.keys()\n",
    "len(variant_sets.get('Enhancer'))\n",
    "len(variant_sets.get(\"5' UTR variant\"))\n",
    "list(variant_sets.get(\"5' UTR variant\"))[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Panel A: *cis*-QTLs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "56200"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "array(['ENSG00000223972.5', 'ENSG00000227232.5', 'ENSG00000278267.1',\n",
       "       'ENSG00000243485.5', 'ENSG00000237613.2'], dtype='<U24')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# annot.genes is a numpy array, I want to take a peek of its content\n",
    "\n",
    "len(annot.gene_ids)\n",
    "annot.gene_ids[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSG00000223972.5 DDX11L1 transcribed_unprocessed_pseudogene chr1 + 11869 14403\n",
      "ENSG00000227232.5 WASH7P unprocessed_pseudogene chr1 - 14410 29553\n",
      "ENSG00000278267.1 MIR6859-1 miRNA chr1 - 17369 17436\n",
      "ENSG00000243485.5 MIR1302-2HG lincRNA chr1 + 29571 31109\n",
      "ENSG00000237613.2 FAM138A lincRNA chr1 - 34554 36081\n"
     ]
    }
   ],
   "source": [
    "for k in annot.gene_ids[:5]:\n",
    "    g = annot.gene_dict[k]\n",
    "    print(g.id, g.name, g.type, g.chr, g.strand, g.start_pos, g.end_pos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# eGenes and sGenes files\n",
    "egene_files = {os.path.basename(i).split('.')[0]:i for i in glob.glob('../data/GTEx_Analysis_v8_eQTL/*.v8.egenes.txt.gz')}\n",
    "sgene_files = {os.path.basename(i).split('.')[0]:i for i in glob.glob('../data/GTEx_Analysis_v8_sQTL/*.v8.sgenes.txt.gz')}\n",
    "tissues = sorted(egene_files.keys())\n",
    "\n",
    "# select a few tissues\n",
    "# tissues = ['Liver', 'Brain_Cortex', 'Kidney_Cortex', 'Whole_Blood']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing Whole_Bloodtine_Terminal_Ileumctionia"
     ]
    }
   ],
   "source": [
    "\n",
    "# annotate top eQTLs and sQTLs with variant categories\n",
    "egene_variant_cat_df = []\n",
    "sgene_variant_cat_df = []\n",
    "for t in tissues:\n",
    "    print('\\rProcessing {}'.format(t), end='')\n",
    "    egenes_df = pd.read_csv(egene_files[t], sep='\\t', index_col=0)\n",
    "    egenes_df['biotype'] = egenes_df.index.map(lambda x: annot.gene_dict[x].type)\n",
    "    egenes_df = egenes_df[(egenes_df['qval']<=0.05) & egenes_df['biotype'].isin(['protein_coding', 'lincRNA'])]\n",
    "    variant_ids = egenes_df['variant_id'].values\n",
    "    pct_s = variant_set_s.apply(lambda x: len([i for i in variant_ids if i in x]))\n",
    "    pct_s.name = t\n",
    "    egene_variant_cat_df.append(pct_s)\n",
    "\n",
    "    sgenes_df = pd.read_csv(sgene_files[t], sep='\\t', index_col=0)\n",
    "    sgenes_df['biotype'] = sgenes_df['gene_id'].apply(lambda x: annot.gene_dict[x].type)\n",
    "    sgenes_df = sgenes_df[(sgenes_df['qval']<=0.05) & sgenes_df['biotype'].isin(['protein_coding', 'lincRNA'])]\n",
    "    variant_ids = sgenes_df['variant_id'].values\n",
    "    pct_s = variant_set_s.apply(lambda x: len([i for i in variant_ids if i in x]))\n",
    "    pct_s.name = t\n",
    "    sgene_variant_cat_df.append(pct_s)\n",
    "    \n",
    "egene_variant_cat_df = pd.concat(egene_variant_cat_df, axis=1)    \n",
    "sgene_variant_cat_df = pd.concat(sgene_variant_cat_df, axis=1)    \n",
    "egene_variant_cat_pct_df = egene_variant_cat_df / egene_variant_cat_df.sum(0)\n",
    "sgene_variant_cat_pct_df = sgene_variant_cat_df / sgene_variant_cat_df.sum(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adipose_Subcutaneous</th>\n",
       "      <th>Adipose_Visceral_Omentum</th>\n",
       "      <th>Adrenal_Gland</th>\n",
       "      <th>Artery_Aorta</th>\n",
       "      <th>Artery_Coronary</th>\n",
       "      <th>Artery_Tibial</th>\n",
       "      <th>Brain_Amygdala</th>\n",
       "      <th>Brain_Anterior_cingulate_cortex_BA24</th>\n",
       "      <th>Brain_Caudate_basal_ganglia</th>\n",
       "      <th>Brain_Cerebellar_Hemisphere</th>\n",
       "      <th>...</th>\n",
       "      <th>Skin_Not_Sun_Exposed_Suprapubic</th>\n",
       "      <th>Skin_Sun_Exposed_Lower_leg</th>\n",
       "      <th>Small_Intestine_Terminal_Ileum</th>\n",
       "      <th>Spleen</th>\n",
       "      <th>Stomach</th>\n",
       "      <th>Testis</th>\n",
       "      <th>Thyroid</th>\n",
       "      <th>Uterus</th>\n",
       "      <th>Vagina</th>\n",
       "      <th>Whole_Blood</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Enhancer</th>\n",
       "      <td>189</td>\n",
       "      <td>137</td>\n",
       "      <td>116</td>\n",
       "      <td>154</td>\n",
       "      <td>51</td>\n",
       "      <td>172</td>\n",
       "      <td>33</td>\n",
       "      <td>34</td>\n",
       "      <td>89</td>\n",
       "      <td>90</td>\n",
       "      <td>...</td>\n",
       "      <td>164</td>\n",
       "      <td>167</td>\n",
       "      <td>61</td>\n",
       "      <td>124</td>\n",
       "      <td>79</td>\n",
       "      <td>164</td>\n",
       "      <td>175</td>\n",
       "      <td>27</td>\n",
       "      <td>25</td>\n",
       "      <td>136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Promoter</th>\n",
       "      <td>1261</td>\n",
       "      <td>1068</td>\n",
       "      <td>630</td>\n",
       "      <td>1053</td>\n",
       "      <td>482</td>\n",
       "      <td>1377</td>\n",
       "      <td>207</td>\n",
       "      <td>362</td>\n",
       "      <td>554</td>\n",
       "      <td>651</td>\n",
       "      <td>...</td>\n",
       "      <td>1207</td>\n",
       "      <td>1341</td>\n",
       "      <td>440</td>\n",
       "      <td>808</td>\n",
       "      <td>745</td>\n",
       "      <td>1084</td>\n",
       "      <td>1337</td>\n",
       "      <td>267</td>\n",
       "      <td>238</td>\n",
       "      <td>1229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Open chromatin region</th>\n",
       "      <td>370</td>\n",
       "      <td>290</td>\n",
       "      <td>168</td>\n",
       "      <td>300</td>\n",
       "      <td>126</td>\n",
       "      <td>386</td>\n",
       "      <td>78</td>\n",
       "      <td>117</td>\n",
       "      <td>178</td>\n",
       "      <td>222</td>\n",
       "      <td>...</td>\n",
       "      <td>378</td>\n",
       "      <td>387</td>\n",
       "      <td>132</td>\n",
       "      <td>203</td>\n",
       "      <td>179</td>\n",
       "      <td>337</td>\n",
       "      <td>387</td>\n",
       "      <td>62</td>\n",
       "      <td>81</td>\n",
       "      <td>254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Promoter-flanking region</th>\n",
       "      <td>675</td>\n",
       "      <td>570</td>\n",
       "      <td>369</td>\n",
       "      <td>575</td>\n",
       "      <td>235</td>\n",
       "      <td>720</td>\n",
       "      <td>146</td>\n",
       "      <td>222</td>\n",
       "      <td>315</td>\n",
       "      <td>361</td>\n",
       "      <td>...</td>\n",
       "      <td>669</td>\n",
       "      <td>710</td>\n",
       "      <td>262</td>\n",
       "      <td>503</td>\n",
       "      <td>351</td>\n",
       "      <td>581</td>\n",
       "      <td>754</td>\n",
       "      <td>123</td>\n",
       "      <td>120</td>\n",
       "      <td>905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTCF binding site</th>\n",
       "      <td>306</td>\n",
       "      <td>258</td>\n",
       "      <td>154</td>\n",
       "      <td>250</td>\n",
       "      <td>124</td>\n",
       "      <td>284</td>\n",
       "      <td>75</td>\n",
       "      <td>128</td>\n",
       "      <td>172</td>\n",
       "      <td>230</td>\n",
       "      <td>...</td>\n",
       "      <td>309</td>\n",
       "      <td>327</td>\n",
       "      <td>133</td>\n",
       "      <td>195</td>\n",
       "      <td>133</td>\n",
       "      <td>371</td>\n",
       "      <td>341</td>\n",
       "      <td>58</td>\n",
       "      <td>77</td>\n",
       "      <td>209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TF binding site</th>\n",
       "      <td>95</td>\n",
       "      <td>90</td>\n",
       "      <td>60</td>\n",
       "      <td>88</td>\n",
       "      <td>43</td>\n",
       "      <td>97</td>\n",
       "      <td>20</td>\n",
       "      <td>39</td>\n",
       "      <td>53</td>\n",
       "      <td>61</td>\n",
       "      <td>...</td>\n",
       "      <td>103</td>\n",
       "      <td>101</td>\n",
       "      <td>49</td>\n",
       "      <td>63</td>\n",
       "      <td>57</td>\n",
       "      <td>149</td>\n",
       "      <td>123</td>\n",
       "      <td>26</td>\n",
       "      <td>26</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3' UTR variant</th>\n",
       "      <td>586</td>\n",
       "      <td>475</td>\n",
       "      <td>276</td>\n",
       "      <td>493</td>\n",
       "      <td>231</td>\n",
       "      <td>662</td>\n",
       "      <td>118</td>\n",
       "      <td>189</td>\n",
       "      <td>334</td>\n",
       "      <td>393</td>\n",
       "      <td>...</td>\n",
       "      <td>604</td>\n",
       "      <td>664</td>\n",
       "      <td>213</td>\n",
       "      <td>382</td>\n",
       "      <td>326</td>\n",
       "      <td>578</td>\n",
       "      <td>655</td>\n",
       "      <td>125</td>\n",
       "      <td>126</td>\n",
       "      <td>421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5' UTR variant</th>\n",
       "      <td>531</td>\n",
       "      <td>437</td>\n",
       "      <td>238</td>\n",
       "      <td>429</td>\n",
       "      <td>199</td>\n",
       "      <td>515</td>\n",
       "      <td>72</td>\n",
       "      <td>135</td>\n",
       "      <td>189</td>\n",
       "      <td>250</td>\n",
       "      <td>...</td>\n",
       "      <td>486</td>\n",
       "      <td>547</td>\n",
       "      <td>189</td>\n",
       "      <td>292</td>\n",
       "      <td>314</td>\n",
       "      <td>464</td>\n",
       "      <td>531</td>\n",
       "      <td>112</td>\n",
       "      <td>102</td>\n",
       "      <td>468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Frameshift variant</th>\n",
       "      <td>13</td>\n",
       "      <td>15</td>\n",
       "      <td>8</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>16</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>14</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "      <td>14</td>\n",
       "      <td>16</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Intron variant</th>\n",
       "      <td>6487</td>\n",
       "      <td>5202</td>\n",
       "      <td>3462</td>\n",
       "      <td>5408</td>\n",
       "      <td>2607</td>\n",
       "      <td>6407</td>\n",
       "      <td>1494</td>\n",
       "      <td>2278</td>\n",
       "      <td>3466</td>\n",
       "      <td>4152</td>\n",
       "      <td>...</td>\n",
       "      <td>6453</td>\n",
       "      <td>7015</td>\n",
       "      <td>2635</td>\n",
       "      <td>4388</td>\n",
       "      <td>3674</td>\n",
       "      <td>7033</td>\n",
       "      <td>7195</td>\n",
       "      <td>1342</td>\n",
       "      <td>1402</td>\n",
       "      <td>5288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Missense variant</th>\n",
       "      <td>236</td>\n",
       "      <td>174</td>\n",
       "      <td>109</td>\n",
       "      <td>185</td>\n",
       "      <td>70</td>\n",
       "      <td>222</td>\n",
       "      <td>67</td>\n",
       "      <td>71</td>\n",
       "      <td>127</td>\n",
       "      <td>144</td>\n",
       "      <td>...</td>\n",
       "      <td>227</td>\n",
       "      <td>255</td>\n",
       "      <td>93</td>\n",
       "      <td>160</td>\n",
       "      <td>133</td>\n",
       "      <td>238</td>\n",
       "      <td>253</td>\n",
       "      <td>39</td>\n",
       "      <td>52</td>\n",
       "      <td>185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Non-coding transcript exon variant</th>\n",
       "      <td>656</td>\n",
       "      <td>568</td>\n",
       "      <td>380</td>\n",
       "      <td>533</td>\n",
       "      <td>264</td>\n",
       "      <td>690</td>\n",
       "      <td>155</td>\n",
       "      <td>236</td>\n",
       "      <td>378</td>\n",
       "      <td>402</td>\n",
       "      <td>...</td>\n",
       "      <td>635</td>\n",
       "      <td>714</td>\n",
       "      <td>293</td>\n",
       "      <td>509</td>\n",
       "      <td>398</td>\n",
       "      <td>835</td>\n",
       "      <td>774</td>\n",
       "      <td>167</td>\n",
       "      <td>165</td>\n",
       "      <td>534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Splice acceptor variant</th>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>19</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Splice donor variant</th>\n",
       "      <td>14</td>\n",
       "      <td>16</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>12</td>\n",
       "      <td>14</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>17</td>\n",
       "      <td>13</td>\n",
       "      <td>17</td>\n",
       "      <td>10</td>\n",
       "      <td>32</td>\n",
       "      <td>21</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Splice region variant</th>\n",
       "      <td>117</td>\n",
       "      <td>87</td>\n",
       "      <td>45</td>\n",
       "      <td>87</td>\n",
       "      <td>45</td>\n",
       "      <td>102</td>\n",
       "      <td>26</td>\n",
       "      <td>35</td>\n",
       "      <td>50</td>\n",
       "      <td>64</td>\n",
       "      <td>...</td>\n",
       "      <td>88</td>\n",
       "      <td>108</td>\n",
       "      <td>46</td>\n",
       "      <td>79</td>\n",
       "      <td>57</td>\n",
       "      <td>106</td>\n",
       "      <td>116</td>\n",
       "      <td>29</td>\n",
       "      <td>28</td>\n",
       "      <td>93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Stop gained</th>\n",
       "      <td>11</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>14</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Synonymous variant</th>\n",
       "      <td>212</td>\n",
       "      <td>149</td>\n",
       "      <td>97</td>\n",
       "      <td>162</td>\n",
       "      <td>91</td>\n",
       "      <td>192</td>\n",
       "      <td>52</td>\n",
       "      <td>64</td>\n",
       "      <td>119</td>\n",
       "      <td>128</td>\n",
       "      <td>...</td>\n",
       "      <td>194</td>\n",
       "      <td>211</td>\n",
       "      <td>81</td>\n",
       "      <td>139</td>\n",
       "      <td>138</td>\n",
       "      <td>191</td>\n",
       "      <td>236</td>\n",
       "      <td>54</td>\n",
       "      <td>39</td>\n",
       "      <td>179</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17 rows × 49 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    Adipose_Subcutaneous  \\\n",
       "Enhancer                                             189   \n",
       "Promoter                                            1261   \n",
       "Open chromatin region                                370   \n",
       "Promoter-flanking region                             675   \n",
       "CTCF binding site                                    306   \n",
       "TF binding site                                       95   \n",
       "3' UTR variant                                       586   \n",
       "5' UTR variant                                       531   \n",
       "Frameshift variant                                    13   \n",
       "Intron variant                                      6487   \n",
       "Missense variant                                     236   \n",
       "Non-coding transcript exon variant                   656   \n",
       "Splice acceptor variant                                8   \n",
       "Splice donor variant                                  14   \n",
       "Splice region variant                                117   \n",
       "Stop gained                                           11   \n",
       "Synonymous variant                                   212   \n",
       "\n",
       "                                    Adipose_Visceral_Omentum  Adrenal_Gland  \\\n",
       "Enhancer                                                 137            116   \n",
       "Promoter                                                1068            630   \n",
       "Open chromatin region                                    290            168   \n",
       "Promoter-flanking region                                 570            369   \n",
       "CTCF binding site                                        258            154   \n",
       "TF binding site                                           90             60   \n",
       "3' UTR variant                                           475            276   \n",
       "5' UTR variant                                           437            238   \n",
       "Frameshift variant                                        15              8   \n",
       "Intron variant                                          5202           3462   \n",
       "Missense variant                                         174            109   \n",
       "Non-coding transcript exon variant                       568            380   \n",
       "Splice acceptor variant                                    7              6   \n",
       "Splice donor variant                                      16             15   \n",
       "Splice region variant                                     87             45   \n",
       "Stop gained                                                9              6   \n",
       "Synonymous variant                                       149             97   \n",
       "\n",
       "                                    Artery_Aorta  Artery_Coronary  \\\n",
       "Enhancer                                     154               51   \n",
       "Promoter                                    1053              482   \n",
       "Open chromatin region                        300              126   \n",
       "Promoter-flanking region                     575              235   \n",
       "CTCF binding site                            250              124   \n",
       "TF binding site                               88               43   \n",
       "3' UTR variant                               493              231   \n",
       "5' UTR variant                               429              199   \n",
       "Frameshift variant                            13                5   \n",
       "Intron variant                              5408             2607   \n",
       "Missense variant                             185               70   \n",
       "Non-coding transcript exon variant           533              264   \n",
       "Splice acceptor variant                       10                5   \n",
       "Splice donor variant                          15               12   \n",
       "Splice region variant                         87               45   \n",
       "Stop gained                                    6                3   \n",
       "Synonymous variant                           162               91   \n",
       "\n",
       "                                    Artery_Tibial  Brain_Amygdala  \\\n",
       "Enhancer                                      172              33   \n",
       "Promoter                                     1377             207   \n",
       "Open chromatin region                         386              78   \n",
       "Promoter-flanking region                      720             146   \n",
       "CTCF binding site                             284              75   \n",
       "TF binding site                                97              20   \n",
       "3' UTR variant                                662             118   \n",
       "5' UTR variant                                515              72   \n",
       "Frameshift variant                             16               5   \n",
       "Intron variant                               6407            1494   \n",
       "Missense variant                              222              67   \n",
       "Non-coding transcript exon variant            690             155   \n",
       "Splice acceptor variant                         9               3   \n",
       "Splice donor variant                           14               4   \n",
       "Splice region variant                         102              26   \n",
       "Stop gained                                     9               1   \n",
       "Synonymous variant                            192              52   \n",
       "\n",
       "                                    Brain_Anterior_cingulate_cortex_BA24  \\\n",
       "Enhancer                                                              34   \n",
       "Promoter                                                             362   \n",
       "Open chromatin region                                                117   \n",
       "Promoter-flanking region                                             222   \n",
       "CTCF binding site                                                    128   \n",
       "TF binding site                                                       39   \n",
       "3' UTR variant                                                       189   \n",
       "5' UTR variant                                                       135   \n",
       "Frameshift variant                                                     5   \n",
       "Intron variant                                                      2278   \n",
       "Missense variant                                                      71   \n",
       "Non-coding transcript exon variant                                   236   \n",
       "Splice acceptor variant                                                4   \n",
       "Splice donor variant                                                   3   \n",
       "Splice region variant                                                 35   \n",
       "Stop gained                                                            2   \n",
       "Synonymous variant                                                    64   \n",
       "\n",
       "                                    Brain_Caudate_basal_ganglia  \\\n",
       "Enhancer                                                     89   \n",
       "Promoter                                                    554   \n",
       "Open chromatin region                                       178   \n",
       "Promoter-flanking region                                    315   \n",
       "CTCF binding site                                           172   \n",
       "TF binding site                                              53   \n",
       "3' UTR variant                                              334   \n",
       "5' UTR variant                                              189   \n",
       "Frameshift variant                                            8   \n",
       "Intron variant                                             3466   \n",
       "Missense variant                                            127   \n",
       "Non-coding transcript exon variant                          378   \n",
       "Splice acceptor variant                                       3   \n",
       "Splice donor variant                                         10   \n",
       "Splice region variant                                        50   \n",
       "Stop gained                                                   5   \n",
       "Synonymous variant                                          119   \n",
       "\n",
       "                                    Brain_Cerebellar_Hemisphere  ...  \\\n",
       "Enhancer                                                     90  ...   \n",
       "Promoter                                                    651  ...   \n",
       "Open chromatin region                                       222  ...   \n",
       "Promoter-flanking region                                    361  ...   \n",
       "CTCF binding site                                           230  ...   \n",
       "TF binding site                                              61  ...   \n",
       "3' UTR variant                                              393  ...   \n",
       "5' UTR variant                                              250  ...   \n",
       "Frameshift variant                                            4  ...   \n",
       "Intron variant                                             4152  ...   \n",
       "Missense variant                                            144  ...   \n",
       "Non-coding transcript exon variant                          402  ...   \n",
       "Splice acceptor variant                                       7  ...   \n",
       "Splice donor variant                                         13  ...   \n",
       "Splice region variant                                        64  ...   \n",
       "Stop gained                                                   2  ...   \n",
       "Synonymous variant                                          128  ...   \n",
       "\n",
       "                                    Skin_Not_Sun_Exposed_Suprapubic  \\\n",
       "Enhancer                                                        164   \n",
       "Promoter                                                       1207   \n",
       "Open chromatin region                                           378   \n",
       "Promoter-flanking region                                        669   \n",
       "CTCF binding site                                               309   \n",
       "TF binding site                                                 103   \n",
       "3' UTR variant                                                  604   \n",
       "5' UTR variant                                                  486   \n",
       "Frameshift variant                                               16   \n",
       "Intron variant                                                 6453   \n",
       "Missense variant                                                227   \n",
       "Non-coding transcript exon variant                              635   \n",
       "Splice acceptor variant                                           8   \n",
       "Splice donor variant                                             15   \n",
       "Splice region variant                                            88   \n",
       "Stop gained                                                      10   \n",
       "Synonymous variant                                              194   \n",
       "\n",
       "                                    Skin_Sun_Exposed_Lower_leg  \\\n",
       "Enhancer                                                   167   \n",
       "Promoter                                                  1341   \n",
       "Open chromatin region                                      387   \n",
       "Promoter-flanking region                                   710   \n",
       "CTCF binding site                                          327   \n",
       "TF binding site                                            101   \n",
       "3' UTR variant                                             664   \n",
       "5' UTR variant                                             547   \n",
       "Frameshift variant                                          14   \n",
       "Intron variant                                            7015   \n",
       "Missense variant                                           255   \n",
       "Non-coding transcript exon variant                         714   \n",
       "Splice acceptor variant                                     12   \n",
       "Splice donor variant                                        17   \n",
       "Splice region variant                                      108   \n",
       "Stop gained                                                 11   \n",
       "Synonymous variant                                         211   \n",
       "\n",
       "                                    Small_Intestine_Terminal_Ileum  Spleen  \\\n",
       "Enhancer                                                        61     124   \n",
       "Promoter                                                       440     808   \n",
       "Open chromatin region                                          132     203   \n",
       "Promoter-flanking region                                       262     503   \n",
       "CTCF binding site                                              133     195   \n",
       "TF binding site                                                 49      63   \n",
       "3' UTR variant                                                 213     382   \n",
       "5' UTR variant                                                 189     292   \n",
       "Frameshift variant                                               3       7   \n",
       "Intron variant                                                2635    4388   \n",
       "Missense variant                                                93     160   \n",
       "Non-coding transcript exon variant                             293     509   \n",
       "Splice acceptor variant                                          5       7   \n",
       "Splice donor variant                                            13      17   \n",
       "Splice region variant                                           46      79   \n",
       "Stop gained                                                      4       4   \n",
       "Synonymous variant                                              81     139   \n",
       "\n",
       "                                    Stomach  Testis  Thyroid  Uterus  Vagina  \\\n",
       "Enhancer                                 79     164      175      27      25   \n",
       "Promoter                                745    1084     1337     267     238   \n",
       "Open chromatin region                   179     337      387      62      81   \n",
       "Promoter-flanking region                351     581      754     123     120   \n",
       "CTCF binding site                       133     371      341      58      77   \n",
       "TF binding site                          57     149      123      26      26   \n",
       "3' UTR variant                          326     578      655     125     126   \n",
       "5' UTR variant                          314     464      531     112     102   \n",
       "Frameshift variant                        9      14       16       3       3   \n",
       "Intron variant                         3674    7033     7195    1342    1402   \n",
       "Missense variant                        133     238      253      39      52   \n",
       "Non-coding transcript exon variant      398     835      774     167     165   \n",
       "Splice acceptor variant                   3      19       11       1       0   \n",
       "Splice donor variant                     10      32       21       6       3   \n",
       "Splice region variant                    57     106      116      29      28   \n",
       "Stop gained                               9      14        7       3       3   \n",
       "Synonymous variant                      138     191      236      54      39   \n",
       "\n",
       "                                    Whole_Blood  \n",
       "Enhancer                                    136  \n",
       "Promoter                                   1229  \n",
       "Open chromatin region                       254  \n",
       "Promoter-flanking region                    905  \n",
       "CTCF binding site                           209  \n",
       "TF binding site                              81  \n",
       "3' UTR variant                              421  \n",
       "5' UTR variant                              468  \n",
       "Frameshift variant                           10  \n",
       "Intron variant                             5288  \n",
       "Missense variant                            185  \n",
       "Non-coding transcript exon variant          534  \n",
       "Splice acceptor variant                       8  \n",
       "Splice donor variant                         11  \n",
       "Splice region variant                        93  \n",
       "Stop gained                                   9  \n",
       "Synonymous variant                          179  \n",
       "\n",
       "[17 rows x 49 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "egene_variant_cat_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Adipose_Subcutaneous        4797\n",
       "Adipose_Visceral_Omentum    3973\n",
       "Adrenal_Gland               2152\n",
       "Artery_Aorta                3434\n",
       "Artery_Coronary             1923\n",
       "dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adipose_Subcutaneous</th>\n",
       "      <th>Adipose_Visceral_Omentum</th>\n",
       "      <th>Adrenal_Gland</th>\n",
       "      <th>Artery_Aorta</th>\n",
       "      <th>Artery_Coronary</th>\n",
       "      <th>Artery_Tibial</th>\n",
       "      <th>Brain_Amygdala</th>\n",
       "      <th>Brain_Anterior_cingulate_cortex_BA24</th>\n",
       "      <th>Brain_Caudate_basal_ganglia</th>\n",
       "      <th>Brain_Cerebellar_Hemisphere</th>\n",
       "      <th>...</th>\n",
       "      <th>Skin_Not_Sun_Exposed_Suprapubic</th>\n",
       "      <th>Skin_Sun_Exposed_Lower_leg</th>\n",
       "      <th>Small_Intestine_Terminal_Ileum</th>\n",
       "      <th>Spleen</th>\n",
       "      <th>Stomach</th>\n",
       "      <th>Testis</th>\n",
       "      <th>Thyroid</th>\n",
       "      <th>Uterus</th>\n",
       "      <th>Vagina</th>\n",
       "      <th>Whole_Blood</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Enhancer</th>\n",
       "      <td>0.007922</td>\n",
       "      <td>0.008809</td>\n",
       "      <td>0.007435</td>\n",
       "      <td>0.007280</td>\n",
       "      <td>0.008320</td>\n",
       "      <td>0.009476</td>\n",
       "      <td>0.011658</td>\n",
       "      <td>0.008588</td>\n",
       "      <td>0.008889</td>\n",
       "      <td>0.007656</td>\n",
       "      <td>...</td>\n",
       "      <td>0.010621</td>\n",
       "      <td>0.008374</td>\n",
       "      <td>0.009179</td>\n",
       "      <td>0.009645</td>\n",
       "      <td>0.009132</td>\n",
       "      <td>0.010125</td>\n",
       "      <td>0.007740</td>\n",
       "      <td>0.008929</td>\n",
       "      <td>0.010980</td>\n",
       "      <td>0.010753</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Promoter</th>\n",
       "      <td>0.068793</td>\n",
       "      <td>0.079789</td>\n",
       "      <td>0.081784</td>\n",
       "      <td>0.066977</td>\n",
       "      <td>0.076443</td>\n",
       "      <td>0.074923</td>\n",
       "      <td>0.080311</td>\n",
       "      <td>0.070611</td>\n",
       "      <td>0.074286</td>\n",
       "      <td>0.070335</td>\n",
       "      <td>...</td>\n",
       "      <td>0.067345</td>\n",
       "      <td>0.067606</td>\n",
       "      <td>0.079374</td>\n",
       "      <td>0.077546</td>\n",
       "      <td>0.078456</td>\n",
       "      <td>0.056189</td>\n",
       "      <td>0.069458</td>\n",
       "      <td>0.077381</td>\n",
       "      <td>0.069804</td>\n",
       "      <td>0.075963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Open chromatin region</th>\n",
       "      <td>0.026892</td>\n",
       "      <td>0.025422</td>\n",
       "      <td>0.025558</td>\n",
       "      <td>0.025917</td>\n",
       "      <td>0.022361</td>\n",
       "      <td>0.027104</td>\n",
       "      <td>0.027202</td>\n",
       "      <td>0.017176</td>\n",
       "      <td>0.017778</td>\n",
       "      <td>0.022010</td>\n",
       "      <td>...</td>\n",
       "      <td>0.025085</td>\n",
       "      <td>0.027982</td>\n",
       "      <td>0.023758</td>\n",
       "      <td>0.022377</td>\n",
       "      <td>0.021586</td>\n",
       "      <td>0.025956</td>\n",
       "      <td>0.024013</td>\n",
       "      <td>0.020833</td>\n",
       "      <td>0.018039</td>\n",
       "      <td>0.024974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Promoter-flanking region</th>\n",
       "      <td>0.042944</td>\n",
       "      <td>0.042789</td>\n",
       "      <td>0.043680</td>\n",
       "      <td>0.039895</td>\n",
       "      <td>0.039002</td>\n",
       "      <td>0.039665</td>\n",
       "      <td>0.041451</td>\n",
       "      <td>0.046756</td>\n",
       "      <td>0.040635</td>\n",
       "      <td>0.043541</td>\n",
       "      <td>...</td>\n",
       "      <td>0.043390</td>\n",
       "      <td>0.043913</td>\n",
       "      <td>0.038337</td>\n",
       "      <td>0.050540</td>\n",
       "      <td>0.035699</td>\n",
       "      <td>0.038791</td>\n",
       "      <td>0.039889</td>\n",
       "      <td>0.040179</td>\n",
       "      <td>0.037647</td>\n",
       "      <td>0.060007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTCF binding site</th>\n",
       "      <td>0.016677</td>\n",
       "      <td>0.015605</td>\n",
       "      <td>0.019981</td>\n",
       "      <td>0.014851</td>\n",
       "      <td>0.014561</td>\n",
       "      <td>0.013222</td>\n",
       "      <td>0.020725</td>\n",
       "      <td>0.021947</td>\n",
       "      <td>0.021587</td>\n",
       "      <td>0.016746</td>\n",
       "      <td>...</td>\n",
       "      <td>0.021017</td>\n",
       "      <td>0.014297</td>\n",
       "      <td>0.012959</td>\n",
       "      <td>0.013117</td>\n",
       "      <td>0.017435</td>\n",
       "      <td>0.020964</td>\n",
       "      <td>0.018059</td>\n",
       "      <td>0.022321</td>\n",
       "      <td>0.019608</td>\n",
       "      <td>0.012487</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 49 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Adipose_Subcutaneous  Adipose_Visceral_Omentum  \\\n",
       "Enhancer                              0.007922                  0.008809   \n",
       "Promoter                              0.068793                  0.079789   \n",
       "Open chromatin region                 0.026892                  0.025422   \n",
       "Promoter-flanking region              0.042944                  0.042789   \n",
       "CTCF binding site                     0.016677                  0.015605   \n",
       "\n",
       "                          Adrenal_Gland  Artery_Aorta  Artery_Coronary  \\\n",
       "Enhancer                       0.007435      0.007280         0.008320   \n",
       "Promoter                       0.081784      0.066977         0.076443   \n",
       "Open chromatin region          0.025558      0.025917         0.022361   \n",
       "Promoter-flanking region       0.043680      0.039895         0.039002   \n",
       "CTCF binding site              0.019981      0.014851         0.014561   \n",
       "\n",
       "                          Artery_Tibial  Brain_Amygdala  \\\n",
       "Enhancer                       0.009476        0.011658   \n",
       "Promoter                       0.074923        0.080311   \n",
       "Open chromatin region          0.027104        0.027202   \n",
       "Promoter-flanking region       0.039665        0.041451   \n",
       "CTCF binding site              0.013222        0.020725   \n",
       "\n",
       "                          Brain_Anterior_cingulate_cortex_BA24  \\\n",
       "Enhancer                                              0.008588   \n",
       "Promoter                                              0.070611   \n",
       "Open chromatin region                                 0.017176   \n",
       "Promoter-flanking region                              0.046756   \n",
       "CTCF binding site                                     0.021947   \n",
       "\n",
       "                          Brain_Caudate_basal_ganglia  \\\n",
       "Enhancer                                     0.008889   \n",
       "Promoter                                     0.074286   \n",
       "Open chromatin region                        0.017778   \n",
       "Promoter-flanking region                     0.040635   \n",
       "CTCF binding site                            0.021587   \n",
       "\n",
       "                          Brain_Cerebellar_Hemisphere  ...  \\\n",
       "Enhancer                                     0.007656  ...   \n",
       "Promoter                                     0.070335  ...   \n",
       "Open chromatin region                        0.022010  ...   \n",
       "Promoter-flanking region                     0.043541  ...   \n",
       "CTCF binding site                            0.016746  ...   \n",
       "\n",
       "                          Skin_Not_Sun_Exposed_Suprapubic  \\\n",
       "Enhancer                                         0.010621   \n",
       "Promoter                                         0.067345   \n",
       "Open chromatin region                            0.025085   \n",
       "Promoter-flanking region                         0.043390   \n",
       "CTCF binding site                                0.021017   \n",
       "\n",
       "                          Skin_Sun_Exposed_Lower_leg  \\\n",
       "Enhancer                                    0.008374   \n",
       "Promoter                                    0.067606   \n",
       "Open chromatin region                       0.027982   \n",
       "Promoter-flanking region                    0.043913   \n",
       "CTCF binding site                           0.014297   \n",
       "\n",
       "                          Small_Intestine_Terminal_Ileum    Spleen   Stomach  \\\n",
       "Enhancer                                        0.009179  0.009645  0.009132   \n",
       "Promoter                                        0.079374  0.077546  0.078456   \n",
       "Open chromatin region                           0.023758  0.022377  0.021586   \n",
       "Promoter-flanking region                        0.038337  0.050540  0.035699   \n",
       "CTCF binding site                               0.012959  0.013117  0.017435   \n",
       "\n",
       "                            Testis   Thyroid    Uterus    Vagina  Whole_Blood  \n",
       "Enhancer                  0.010125  0.007740  0.008929  0.010980     0.010753  \n",
       "Promoter                  0.056189  0.069458  0.077381  0.069804     0.075963  \n",
       "Open chromatin region     0.025956  0.024013  0.020833  0.018039     0.024974  \n",
       "Promoter-flanking region  0.038791  0.039889  0.040179  0.037647     0.060007  \n",
       "CTCF binding site         0.020964  0.018059  0.022321  0.019608     0.012487  \n",
       "\n",
       "[5 rows x 49 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sgene_variant_cat_df.sum(0).head()\n",
    "sgene_variant_cat_pct_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['gene_name', 'gene_chr', 'gene_start', 'gene_end', 'strand', 'num_var',\n",
       "       'beta_shape1', 'beta_shape2', 'true_df', 'pval_true_df', 'variant_id',\n",
       "       'tss_distance', 'chr', 'variant_pos', 'ref', 'alt', 'num_alt_per_site',\n",
       "       'rs_id_dbSNP151_GRCh38p7', 'minor_allele_samples', 'minor_allele_count',\n",
       "       'maf', 'ref_factor', 'pval_nominal', 'slope', 'slope_se', 'pval_perm',\n",
       "       'pval_beta', 'qval', 'pval_nominal_threshold', 'log2_aFC',\n",
       "       'log2_aFC_lower', 'log2_aFC_upper', 'biotype'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_chr</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>num_var</th>\n",
       "      <th>beta_shape1</th>\n",
       "      <th>beta_shape2</th>\n",
       "      <th>true_df</th>\n",
       "      <th>pval_true_df</th>\n",
       "      <th>...</th>\n",
       "      <th>slope</th>\n",
       "      <th>slope_se</th>\n",
       "      <th>pval_perm</th>\n",
       "      <th>pval_beta</th>\n",
       "      <th>qval</th>\n",
       "      <th>pval_nominal_threshold</th>\n",
       "      <th>log2_aFC</th>\n",
       "      <th>log2_aFC_lower</th>\n",
       "      <th>log2_aFC_upper</th>\n",
       "      <th>biotype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009.6</th>\n",
       "      <td>RP11-34P13.7</td>\n",
       "      <td>chr1</td>\n",
       "      <td>89295</td>\n",
       "      <td>129223</td>\n",
       "      <td>-</td>\n",
       "      <td>1807</td>\n",
       "      <td>1.01210</td>\n",
       "      <td>354.630</td>\n",
       "      <td>537.158</td>\n",
       "      <td>5.974450e-28</td>\n",
       "      <td>...</td>\n",
       "      <td>0.693059</td>\n",
       "      <td>0.056494</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>1.059730e-25</td>\n",
       "      <td>3.024780e-25</td>\n",
       "      <td>0.000333</td>\n",
       "      <td>2.298749</td>\n",
       "      <td>2.013133</td>\n",
       "      <td>2.547666</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000228463.9</th>\n",
       "      <td>AP006222.2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>257864</td>\n",
       "      <td>297502</td>\n",
       "      <td>-</td>\n",
       "      <td>2799</td>\n",
       "      <td>1.03450</td>\n",
       "      <td>466.874</td>\n",
       "      <td>522.949</td>\n",
       "      <td>1.932970e-04</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.533788</td>\n",
       "      <td>0.132736</td>\n",
       "      <td>0.075493</td>\n",
       "      <td>7.819130e-02</td>\n",
       "      <td>3.787400e-02</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>-2.267305</td>\n",
       "      <td>-6.643856</td>\n",
       "      <td>-1.033481</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237491.8</th>\n",
       "      <td>RP11-206L10.9</td>\n",
       "      <td>chr1</td>\n",
       "      <td>778770</td>\n",
       "      <td>810060</td>\n",
       "      <td>+</td>\n",
       "      <td>4834</td>\n",
       "      <td>1.01838</td>\n",
       "      <td>759.415</td>\n",
       "      <td>530.640</td>\n",
       "      <td>1.468930e-23</td>\n",
       "      <td>...</td>\n",
       "      <td>1.026590</td>\n",
       "      <td>0.091948</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>4.757680e-21</td>\n",
       "      <td>1.099550e-20</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>1.450466</td>\n",
       "      <td>1.255606</td>\n",
       "      <td>1.626524</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000177757.2</th>\n",
       "      <td>FAM87B</td>\n",
       "      <td>chr1</td>\n",
       "      <td>817371</td>\n",
       "      <td>819837</td>\n",
       "      <td>+</td>\n",
       "      <td>4893</td>\n",
       "      <td>1.05708</td>\n",
       "      <td>823.141</td>\n",
       "      <td>535.312</td>\n",
       "      <td>1.160180e-06</td>\n",
       "      <td>...</td>\n",
       "      <td>0.627772</td>\n",
       "      <td>0.120550</td>\n",
       "      <td>0.000400</td>\n",
       "      <td>6.258900e-04</td>\n",
       "      <td>4.408980e-04</td>\n",
       "      <td>0.000162</td>\n",
       "      <td>1.080225</td>\n",
       "      <td>0.766324</td>\n",
       "      <td>1.366386</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000225880.5</th>\n",
       "      <td>LINC00115</td>\n",
       "      <td>chr1</td>\n",
       "      <td>826206</td>\n",
       "      <td>827522</td>\n",
       "      <td>-</td>\n",
       "      <td>4907</td>\n",
       "      <td>1.04771</td>\n",
       "      <td>766.544</td>\n",
       "      <td>524.692</td>\n",
       "      <td>3.418240e-27</td>\n",
       "      <td>...</td>\n",
       "      <td>0.780427</td>\n",
       "      <td>0.063819</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>1.924370e-25</td>\n",
       "      <td>5.441130e-25</td>\n",
       "      <td>0.000169</td>\n",
       "      <td>1.132230</td>\n",
       "      <td>0.999367</td>\n",
       "      <td>1.282875</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       gene_name gene_chr  gene_start  gene_end strand  \\\n",
       "gene_id                                                                  \n",
       "ENSG00000238009.6   RP11-34P13.7     chr1       89295    129223      -   \n",
       "ENSG00000228463.9     AP006222.2     chr1      257864    297502      -   \n",
       "ENSG00000237491.8  RP11-206L10.9     chr1      778770    810060      +   \n",
       "ENSG00000177757.2         FAM87B     chr1      817371    819837      +   \n",
       "ENSG00000225880.5      LINC00115     chr1      826206    827522      -   \n",
       "\n",
       "                   num_var  beta_shape1  beta_shape2  true_df  pval_true_df  \\\n",
       "gene_id                                                                       \n",
       "ENSG00000238009.6     1807      1.01210      354.630  537.158  5.974450e-28   \n",
       "ENSG00000228463.9     2799      1.03450      466.874  522.949  1.932970e-04   \n",
       "ENSG00000237491.8     4834      1.01838      759.415  530.640  1.468930e-23   \n",
       "ENSG00000177757.2     4893      1.05708      823.141  535.312  1.160180e-06   \n",
       "ENSG00000225880.5     4907      1.04771      766.544  524.692  3.418240e-27   \n",
       "\n",
       "                   ...     slope  slope_se pval_perm     pval_beta  \\\n",
       "gene_id            ...                                               \n",
       "ENSG00000238009.6  ...  0.693059  0.056494  0.000100  1.059730e-25   \n",
       "ENSG00000228463.9  ... -0.533788  0.132736  0.075493  7.819130e-02   \n",
       "ENSG00000237491.8  ...  1.026590  0.091948  0.000100  4.757680e-21   \n",
       "ENSG00000177757.2  ...  0.627772  0.120550  0.000400  6.258900e-04   \n",
       "ENSG00000225880.5  ...  0.780427  0.063819  0.000100  1.924370e-25   \n",
       "\n",
       "                           qval pval_nominal_threshold  log2_aFC  \\\n",
       "gene_id                                                            \n",
       "ENSG00000238009.6  3.024780e-25               0.000333  2.298749   \n",
       "ENSG00000228463.9  3.787400e-02               0.000269 -2.267305   \n",
       "ENSG00000237491.8  1.099550e-20               0.000158  1.450466   \n",
       "ENSG00000177757.2  4.408980e-04               0.000162  1.080225   \n",
       "ENSG00000225880.5  5.441130e-25               0.000169  1.132230   \n",
       "\n",
       "                  log2_aFC_lower  log2_aFC_upper  biotype  \n",
       "gene_id                                                    \n",
       "ENSG00000238009.6       2.013133        2.547666  lincRNA  \n",
       "ENSG00000228463.9      -6.643856       -1.033481  lincRNA  \n",
       "ENSG00000237491.8       1.255606        1.626524  lincRNA  \n",
       "ENSG00000177757.2       0.766324        1.366386  lincRNA  \n",
       "ENSG00000225880.5       0.999367        1.282875  lincRNA  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "egenes_df.columns\n",
    "egenes_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "# mean and s.d. of variant categories across tissues\n",
    "mean_df = pd.concat([\n",
    "    egene_variant_cat_pct_df.mean(1),\n",
    "    sgene_variant_cat_pct_df.mean(1)\n",
    "], axis=1)\n",
    "mean_df.columns = ['eQTL', 'sQTL']\n",
    "mean_df.rename(index=qtl.torus.torus_short_dict, inplace =True)\n",
    "\n",
    "std_df = pd.concat([\n",
    "    egene_variant_cat_pct_df.std(1),\n",
    "    sgene_variant_cat_pct_df.std(1)\n",
    "], axis=1)\n",
    "std_df.columns = ['eQTL', 'sQTL']\n",
    "std_df.rename(index=qtl.torus.torus_short_dict, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(17, 3)"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(17, 3)"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sqtl_tissue_avg_df.shape\n",
    "eqtl_tissue_avg_df.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load Torus outputs for eQTLs and sQTLs\n",
    "eqtl_lor_df = qtl.torus.load_summary('../code/workflow/submodules/gtex-v8/data/GTEx_Analysis_v8_eQTL.torus_enrichment.txt', log2=True)\n",
    "sqtl_lor_df = qtl.torus.load_summary('../code/workflow/submodules/gtex-v8/data/GTEx_Analysis_v8_sQTL.torus_enrichment.txt', log2=True)\n",
    "eqtl_lor_df.drop(['Kidney_Cortex'], axis=1, inplace=True)\n",
    "sqtl_lor_df.drop(['Kidney_Cortex'], axis=1, inplace=True)\n",
    "\n",
    "# average across tissues: cis-eQTLs\n",
    "m = (2**eqtl_lor_df).mean(1)\n",
    "s = (2**eqtl_lor_df).std(1)\n",
    "eqtl_tissue_avg_df = np.log2(pd.concat([m, m-s, m+s], axis=1))\n",
    "eqtl_tissue_avg_df.columns = ['mean', 'std_low', 'std_high']\n",
    "eqtl_tissue_avg_df.rename(index=qtl.torus.torus_short_dict, inplace=True)\n",
    "\n",
    "# average across tissues: cis-sQTLs\n",
    "m = (2**sqtl_lor_df).mean(1)\n",
    "s = (2**sqtl_lor_df).std(1)\n",
    "sqtl_tissue_avg_df = np.log2(pd.concat([m, m-s, m+s], axis=1))\n",
    "sqtl_tissue_avg_df.columns = ['mean', 'std_low', 'std_high']\n",
    "sqtl_tissue_avg_df.rename(index=qtl.torus.torus_short_dict, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adipose_Subcutaneous</th>\n",
       "      <th>Adipose_Visceral_Omentum</th>\n",
       "      <th>Adrenal_Gland</th>\n",
       "      <th>Artery_Aorta</th>\n",
       "      <th>Artery_Coronary</th>\n",
       "      <th>Artery_Tibial</th>\n",
       "      <th>Brain_Amygdala</th>\n",
       "      <th>Brain_Anterior_cingulate_cortex_BA24</th>\n",
       "      <th>Brain_Caudate_basal_ganglia</th>\n",
       "      <th>Brain_Cerebellar_Hemisphere</th>\n",
       "      <th>...</th>\n",
       "      <th>Skin_Not_Sun_Exposed_Suprapubic</th>\n",
       "      <th>Skin_Sun_Exposed_Lower_leg</th>\n",
       "      <th>Small_Intestine_Terminal_Ileum</th>\n",
       "      <th>Spleen</th>\n",
       "      <th>Stomach</th>\n",
       "      <th>Testis</th>\n",
       "      <th>Thyroid</th>\n",
       "      <th>Uterus</th>\n",
       "      <th>Vagina</th>\n",
       "      <th>Whole_Blood</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Enhancer</th>\n",
       "      <td>1.359019</td>\n",
       "      <td>1.198880</td>\n",
       "      <td>1.507616</td>\n",
       "      <td>1.292655</td>\n",
       "      <td>1.015657</td>\n",
       "      <td>1.240718</td>\n",
       "      <td>0.377986</td>\n",
       "      <td>0.288539</td>\n",
       "      <td>0.561208</td>\n",
       "      <td>0.243815</td>\n",
       "      <td>...</td>\n",
       "      <td>0.805024</td>\n",
       "      <td>0.805024</td>\n",
       "      <td>0.536683</td>\n",
       "      <td>1.138286</td>\n",
       "      <td>0.686723</td>\n",
       "      <td>0.699707</td>\n",
       "      <td>0.995460</td>\n",
       "      <td>0.451564</td>\n",
       "      <td>-0.220732</td>\n",
       "      <td>1.346034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Promoter</th>\n",
       "      <td>1.383545</td>\n",
       "      <td>1.374888</td>\n",
       "      <td>1.327279</td>\n",
       "      <td>1.425383</td>\n",
       "      <td>1.292655</td>\n",
       "      <td>1.543684</td>\n",
       "      <td>0.880044</td>\n",
       "      <td>0.920439</td>\n",
       "      <td>1.021428</td>\n",
       "      <td>1.206093</td>\n",
       "      <td>...</td>\n",
       "      <td>1.312852</td>\n",
       "      <td>1.301311</td>\n",
       "      <td>1.112318</td>\n",
       "      <td>1.335936</td>\n",
       "      <td>1.289769</td>\n",
       "      <td>0.747316</td>\n",
       "      <td>1.217635</td>\n",
       "      <td>1.227733</td>\n",
       "      <td>1.133958</td>\n",
       "      <td>1.625917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Open chromatin region</th>\n",
       "      <td>0.644885</td>\n",
       "      <td>0.565536</td>\n",
       "      <td>0.431366</td>\n",
       "      <td>0.755972</td>\n",
       "      <td>0.227946</td>\n",
       "      <td>0.719905</td>\n",
       "      <td>-0.236602</td>\n",
       "      <td>0.036067</td>\n",
       "      <td>0.318836</td>\n",
       "      <td>0.238045</td>\n",
       "      <td>...</td>\n",
       "      <td>0.561208</td>\n",
       "      <td>0.561208</td>\n",
       "      <td>-0.043281</td>\n",
       "      <td>0.154368</td>\n",
       "      <td>0.092332</td>\n",
       "      <td>0.096661</td>\n",
       "      <td>0.328934</td>\n",
       "      <td>-0.217847</td>\n",
       "      <td>-0.432809</td>\n",
       "      <td>0.350575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Promoter-flanking region</th>\n",
       "      <td>0.721348</td>\n",
       "      <td>0.786269</td>\n",
       "      <td>0.790597</td>\n",
       "      <td>0.758858</td>\n",
       "      <td>0.432809</td>\n",
       "      <td>0.764628</td>\n",
       "      <td>0.380871</td>\n",
       "      <td>0.264013</td>\n",
       "      <td>0.369330</td>\n",
       "      <td>0.281326</td>\n",
       "      <td>...</td>\n",
       "      <td>0.620359</td>\n",
       "      <td>0.680952</td>\n",
       "      <td>0.450121</td>\n",
       "      <td>0.890143</td>\n",
       "      <td>0.497730</td>\n",
       "      <td>0.098103</td>\n",
       "      <td>0.637671</td>\n",
       "      <td>0.311622</td>\n",
       "      <td>0.255357</td>\n",
       "      <td>1.651886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTCF binding site</th>\n",
       "      <td>-0.594390</td>\n",
       "      <td>-0.476089</td>\n",
       "      <td>-0.709806</td>\n",
       "      <td>-0.504943</td>\n",
       "      <td>-0.259685</td>\n",
       "      <td>-0.471761</td>\n",
       "      <td>-0.204863</td>\n",
       "      <td>0.178894</td>\n",
       "      <td>-0.541011</td>\n",
       "      <td>-0.429923</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.344804</td>\n",
       "      <td>-0.712691</td>\n",
       "      <td>-0.382314</td>\n",
       "      <td>-0.607375</td>\n",
       "      <td>-0.644885</td>\n",
       "      <td>-0.232274</td>\n",
       "      <td>-0.435694</td>\n",
       "      <td>-0.497730</td>\n",
       "      <td>-0.487631</td>\n",
       "      <td>-0.770399</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Adipose_Subcutaneous  Adipose_Visceral_Omentum  \\\n",
       "feature_id                                                                 \n",
       "Enhancer                              1.359019                  1.198880   \n",
       "Promoter                              1.383545                  1.374888   \n",
       "Open chromatin region                 0.644885                  0.565536   \n",
       "Promoter-flanking region              0.721348                  0.786269   \n",
       "CTCF binding site                    -0.594390                 -0.476089   \n",
       "\n",
       "                          Adrenal_Gland  Artery_Aorta  Artery_Coronary  \\\n",
       "feature_id                                                               \n",
       "Enhancer                       1.507616      1.292655         1.015657   \n",
       "Promoter                       1.327279      1.425383         1.292655   \n",
       "Open chromatin region          0.431366      0.755972         0.227946   \n",
       "Promoter-flanking region       0.790597      0.758858         0.432809   \n",
       "CTCF binding site             -0.709806     -0.504943        -0.259685   \n",
       "\n",
       "                          Artery_Tibial  Brain_Amygdala  \\\n",
       "feature_id                                                \n",
       "Enhancer                       1.240718        0.377986   \n",
       "Promoter                       1.543684        0.880044   \n",
       "Open chromatin region          0.719905       -0.236602   \n",
       "Promoter-flanking region       0.764628        0.380871   \n",
       "CTCF binding site             -0.471761       -0.204863   \n",
       "\n",
       "                          Brain_Anterior_cingulate_cortex_BA24  \\\n",
       "feature_id                                                       \n",
       "Enhancer                                              0.288539   \n",
       "Promoter                                              0.920439   \n",
       "Open chromatin region                                 0.036067   \n",
       "Promoter-flanking region                              0.264013   \n",
       "CTCF binding site                                     0.178894   \n",
       "\n",
       "                          Brain_Caudate_basal_ganglia  \\\n",
       "feature_id                                              \n",
       "Enhancer                                     0.561208   \n",
       "Promoter                                     1.021428   \n",
       "Open chromatin region                        0.318836   \n",
       "Promoter-flanking region                     0.369330   \n",
       "CTCF binding site                           -0.541011   \n",
       "\n",
       "                          Brain_Cerebellar_Hemisphere  ...  \\\n",
       "feature_id                                             ...   \n",
       "Enhancer                                     0.243815  ...   \n",
       "Promoter                                     1.206093  ...   \n",
       "Open chromatin region                        0.238045  ...   \n",
       "Promoter-flanking region                     0.281326  ...   \n",
       "CTCF binding site                           -0.429923  ...   \n",
       "\n",
       "                          Skin_Not_Sun_Exposed_Suprapubic  \\\n",
       "feature_id                                                  \n",
       "Enhancer                                         0.805024   \n",
       "Promoter                                         1.312852   \n",
       "Open chromatin region                            0.561208   \n",
       "Promoter-flanking region                         0.620359   \n",
       "CTCF binding site                               -0.344804   \n",
       "\n",
       "                          Skin_Sun_Exposed_Lower_leg  \\\n",
       "feature_id                                             \n",
       "Enhancer                                    0.805024   \n",
       "Promoter                                    1.301311   \n",
       "Open chromatin region                       0.561208   \n",
       "Promoter-flanking region                    0.680952   \n",
       "CTCF binding site                          -0.712691   \n",
       "\n",
       "                          Small_Intestine_Terminal_Ileum    Spleen   Stomach  \\\n",
       "feature_id                                                                     \n",
       "Enhancer                                        0.536683  1.138286  0.686723   \n",
       "Promoter                                        1.112318  1.335936  1.289769   \n",
       "Open chromatin region                          -0.043281  0.154368  0.092332   \n",
       "Promoter-flanking region                        0.450121  0.890143  0.497730   \n",
       "CTCF binding site                              -0.382314 -0.607375 -0.644885   \n",
       "\n",
       "                            Testis   Thyroid    Uterus    Vagina  Whole_Blood  \n",
       "feature_id                                                                     \n",
       "Enhancer                  0.699707  0.995460  0.451564 -0.220732     1.346034  \n",
       "Promoter                  0.747316  1.217635  1.227733  1.133958     1.625917  \n",
       "Open chromatin region     0.096661  0.328934 -0.217847 -0.432809     0.350575  \n",
       "Promoter-flanking region  0.098103  0.637671  0.311622  0.255357     1.651886  \n",
       "CTCF binding site        -0.232274 -0.435694 -0.497730 -0.487631    -0.770399  \n",
       "\n",
       "[5 rows x 48 columns]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mean</th>\n",
       "      <th>std_low</th>\n",
       "      <th>std_high</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Enhancer</th>\n",
       "      <td>0.310085</td>\n",
       "      <td>-0.003145</td>\n",
       "      <td>0.567292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Promoter</th>\n",
       "      <td>-0.285583</td>\n",
       "      <td>-0.397825</td>\n",
       "      <td>-0.181447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Open chromatin</th>\n",
       "      <td>-0.244861</td>\n",
       "      <td>-0.487409</td>\n",
       "      <td>-0.037282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Promoter-flanking</th>\n",
       "      <td>-0.332649</td>\n",
       "      <td>-0.550822</td>\n",
       "      <td>-0.143177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTCF binding site</th>\n",
       "      <td>-1.054247</td>\n",
       "      <td>-1.680616</td>\n",
       "      <td>-0.618943</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       mean   std_low  std_high\n",
       "feature_id                                     \n",
       "Enhancer           0.310085 -0.003145  0.567292\n",
       "Promoter          -0.285583 -0.397825 -0.181447\n",
       "Open chromatin    -0.244861 -0.487409 -0.037282\n",
       "Promoter-flanking -0.332649 -0.550822 -0.143177\n",
       "CTCF binding site -1.054247 -1.680616 -0.618943"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eQTL</th>\n",
       "      <th>sQTL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Enhancer</th>\n",
       "      <td>0.013273</td>\n",
       "      <td>0.008852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Promoter</th>\n",
       "      <td>0.103287</td>\n",
       "      <td>0.073607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Open chromatin</th>\n",
       "      <td>0.029120</td>\n",
       "      <td>0.024199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Promoter-flanking</th>\n",
       "      <td>0.055833</td>\n",
       "      <td>0.041868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTCF binding site</th>\n",
       "      <td>0.026637</td>\n",
       "      <td>0.017546</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       eQTL      sQTL\n",
       "Enhancer           0.013273  0.008852\n",
       "Promoter           0.103287  0.073607\n",
       "Open chromatin     0.029120  0.024199\n",
       "Promoter-flanking  0.055833  0.041868\n",
       "CTCF binding site  0.026637  0.017546"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# torus\n",
    "eqtl_lor_df.head()\n",
    "\n",
    "sqtl_tissue_avg_df.head() # log2 enrichment\n",
    "mean_df.head() # proportion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (4,) + inhomogeneous part.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[67], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m eargs \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfmt\u001b[39m\u001b[38;5;124m'\u001b[39m:\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m.\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mms\u001b[39m\u001b[38;5;124m'\u001b[39m:\u001b[38;5;241m8\u001b[39m,   \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlw\u001b[39m\u001b[38;5;124m'\u001b[39m:\u001b[38;5;241m1\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mclip_on\u001b[39m\u001b[38;5;124m'\u001b[39m:\u001b[38;5;28;01mFalse\u001b[39;00m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mc\u001b[39m\u001b[38;5;124m'\u001b[39m:style\u001b[38;5;241m.\u001b[39mqtl_color_s[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124meqtl\u001b[39m\u001b[38;5;124m'\u001b[39m], \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlabel\u001b[39m\u001b[38;5;124m'\u001b[39m:\u001b[38;5;124m'\u001b[39m\u001b[38;5;124meQTL\u001b[39m\u001b[38;5;124m'\u001b[39m}\n\u001b[1;32m      2\u001b[0m sargs \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfmt\u001b[39m\u001b[38;5;124m'\u001b[39m:\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mD\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mms\u001b[39m\u001b[38;5;124m'\u001b[39m:\u001b[38;5;241m3.5\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlw\u001b[39m\u001b[38;5;124m'\u001b[39m:\u001b[38;5;241m1\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mclip_on\u001b[39m\u001b[38;5;124m'\u001b[39m:\u001b[38;5;28;01mFalse\u001b[39;00m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mc\u001b[39m\u001b[38;5;124m'\u001b[39m:style\u001b[38;5;241m.\u001b[39mqtl_color_s[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124msqtl\u001b[39m\u001b[38;5;124m'\u001b[39m], \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlabel\u001b[39m\u001b[38;5;124m'\u001b[39m:\u001b[38;5;124m'\u001b[39m\u001b[38;5;124msQTL\u001b[39m\u001b[38;5;124m'\u001b[39m}\n\u001b[0;32m----> 3\u001b[0m \u001b[43mplot_enrichment\u001b[49m\u001b[43m(\u001b[49m\u001b[43meqtl_tissue_avg_df\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msqtl_tissue_avg_df\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmean_df\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstd_df\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43meargs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      4\u001b[0m  \u001b[38;5;66;03m# plt.savefig('pdfs/Fig4A_torus_enrichment_summary.pdf')\u001b[39;00m\n",
      "Cell \u001b[0;32mIn[57], line 22\u001b[0m, in \u001b[0;36mplot_enrichment\u001b[0;34m(df1, df2, varmean_df, varstd_df, args1, args2, legend_title)\u001b[0m\n\u001b[1;32m     19\u001b[0m y \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39marange(df1\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m0\u001b[39m])\n\u001b[1;32m     20\u001b[0m ax1\u001b[38;5;241m.\u001b[39mset_ylim([y[\u001b[38;5;241m0\u001b[39m]\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m0.5\u001b[39m, y[\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m]\u001b[38;5;241m+\u001b[39m\u001b[38;5;241m0.5\u001b[39m])\n\u001b[0;32m---> 22\u001b[0m \u001b[43mqtl\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mplot\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mplot_effects\u001b[49m\u001b[43m(\u001b[49m\u001b[43m[\u001b[49m\u001b[43mdf1\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdf2\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m[\u001b[49m\u001b[43margs1\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43margs2\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     23\u001b[0m \u001b[43m             \u001b[49m\u001b[43mxspace\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m1.25\u001b[39;49m\u001b[43m,\u001b[49m\u001b[38;5;241;43m2\u001b[39;49m\u001b[43m,\u001b[49m\u001b[38;5;241;43m0.5\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43myspace\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0.5\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43mah\u001b[49m\u001b[43m,\u001b[49m\u001b[38;5;241;43m0.25\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mxlim\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m-\u001b[39;49m\u001b[38;5;241;43m2\u001b[39;49m\u001b[43m,\u001b[49m\u001b[38;5;241;43m9\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     24\u001b[0m \u001b[43m             \u001b[49m\u001b[43mxlabel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mlog$\u001b[39;49m\u001b[38;5;124;43m\\\u001b[39;49m\u001b[38;5;124;43mmathregular\u001b[39;49m\u001b[38;5;124;43m{\u001b[39;49m\u001b[38;5;124;43m_\u001b[39;49m\u001b[38;5;132;43;01m{2}\u001b[39;49;00m\u001b[38;5;124;43m}$(Fold enrichment)\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mylabel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43max\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43max1\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     26\u001b[0m ax1\u001b[38;5;241m.\u001b[39mset_yticks(y)\n\u001b[1;32m     27\u001b[0m ax1\u001b[38;5;241m.\u001b[39mset_yticklabels(df1\u001b[38;5;241m.\u001b[39mindex)\n",
      "File \u001b[0;32m/scratch/midway3/chaodai/miniconda3/envs/sos/lib/python3.9/site-packages/qtl/plot.py:412\u001b[0m, in \u001b[0;36mplot_effects\u001b[0;34m(dfs, args, ax, xspace, yspace, xlim, xlabel, ylabel)\u001b[0m\n\u001b[1;32m    410\u001b[0m         c \u001b[38;5;241m=\u001b[39m [\u001b[38;5;241m0.9\u001b[39m]\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m3\u001b[39m\n\u001b[1;32m    411\u001b[0m     patch \u001b[38;5;241m=\u001b[39m patches\u001b[38;5;241m.\u001b[39mRectangle((xlim[\u001b[38;5;241m0\u001b[39m], i\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m0.5\u001b[39m), np\u001b[38;5;241m.\u001b[39mdiff(xlim), \u001b[38;5;241m1\u001b[39m, fc\u001b[38;5;241m=\u001b[39mc, zorder\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m10\u001b[39m)\n\u001b[0;32m--> 412\u001b[0m     \u001b[43max\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43madd_patch\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpatch\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    414\u001b[0m ax\u001b[38;5;241m.\u001b[39mset_xlabel(xlabel, fontsize\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m12\u001b[39m)\n\u001b[1;32m    415\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m ylabel \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m/scratch/midway3/chaodai/miniconda3/envs/sos/lib/python3.9/site-packages/matplotlib/axes/_base.py:2415\u001b[0m, in \u001b[0;36m_AxesBase.add_patch\u001b[0;34m(self, p)\u001b[0m\n\u001b[1;32m   2413\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m p\u001b[38;5;241m.\u001b[39mget_clip_path() \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   2414\u001b[0m     p\u001b[38;5;241m.\u001b[39mset_clip_path(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpatch)\n\u001b[0;32m-> 2415\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_update_patch_limits\u001b[49m\u001b[43m(\u001b[49m\u001b[43mp\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2416\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_children\u001b[38;5;241m.\u001b[39mappend(p)\n\u001b[1;32m   2417\u001b[0m p\u001b[38;5;241m.\u001b[39m_remove_method \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_children\u001b[38;5;241m.\u001b[39mremove\n",
      "File \u001b[0;32m/scratch/midway3/chaodai/miniconda3/envs/sos/lib/python3.9/site-packages/matplotlib/axes/_base.py:2446\u001b[0m, in \u001b[0;36m_AxesBase._update_patch_limits\u001b[0;34m(self, patch)\u001b[0m\n\u001b[1;32m   2443\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(vertices):\n\u001b[1;32m   2444\u001b[0m     vertices \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mrow_stack(vertices)\n\u001b[0;32m-> 2446\u001b[0m patch_trf \u001b[38;5;241m=\u001b[39m \u001b[43mpatch\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_transform\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2447\u001b[0m updatex, updatey \u001b[38;5;241m=\u001b[39m patch_trf\u001b[38;5;241m.\u001b[39mcontains_branch_seperately(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtransData)\n\u001b[1;32m   2448\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (updatex \u001b[38;5;129;01mor\u001b[39;00m updatey):\n",
      "File \u001b[0;32m/scratch/midway3/chaodai/miniconda3/envs/sos/lib/python3.9/site-packages/matplotlib/patches.py:262\u001b[0m, in \u001b[0;36mPatch.get_transform\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    260\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_transform\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m    261\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Return the `~.transforms.Transform` applied to the `Patch`.\"\"\"\u001b[39;00m\n\u001b[0;32m--> 262\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_patch_transform\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;241m+\u001b[39m artist\u001b[38;5;241m.\u001b[39mArtist\u001b[38;5;241m.\u001b[39mget_transform(\u001b[38;5;28mself\u001b[39m)\n",
      "File \u001b[0;32m/scratch/midway3/chaodai/miniconda3/envs/sos/lib/python3.9/site-packages/matplotlib/patches.py:745\u001b[0m, in \u001b[0;36mRectangle.get_patch_transform\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    740\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_patch_transform\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m    741\u001b[0m     \u001b[38;5;66;03m# Note: This cannot be called until after this has been added to\u001b[39;00m\n\u001b[1;32m    742\u001b[0m     \u001b[38;5;66;03m# an Axes, otherwise unit conversion will fail. This makes it very\u001b[39;00m\n\u001b[1;32m    743\u001b[0m     \u001b[38;5;66;03m# important to call the accessor method and not directly access the\u001b[39;00m\n\u001b[1;32m    744\u001b[0m     \u001b[38;5;66;03m# transformation member variable.\u001b[39;00m\n\u001b[0;32m--> 745\u001b[0m     bbox \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_bbox\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    746\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrotation_point \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcenter\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[1;32m    747\u001b[0m         width, height \u001b[38;5;241m=\u001b[39m bbox\u001b[38;5;241m.\u001b[39mx1 \u001b[38;5;241m-\u001b[39m bbox\u001b[38;5;241m.\u001b[39mx0, bbox\u001b[38;5;241m.\u001b[39my1 \u001b[38;5;241m-\u001b[39m bbox\u001b[38;5;241m.\u001b[39my0\n",
      "File \u001b[0;32m/scratch/midway3/chaodai/miniconda3/envs/sos/lib/python3.9/site-packages/matplotlib/patches.py:877\u001b[0m, in \u001b[0;36mRectangle.get_bbox\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    875\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Return the `.Bbox`.\"\"\"\u001b[39;00m\n\u001b[1;32m    876\u001b[0m x0, y0, x1, y1 \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_convert_units()\n\u001b[0;32m--> 877\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mtransforms\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mBbox\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfrom_extents\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx0\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43my0\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mx1\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43my1\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/scratch/midway3/chaodai/miniconda3/envs/sos/lib/python3.9/site-packages/matplotlib/transforms.py:826\u001b[0m, in \u001b[0;36mBbox.from_extents\u001b[0;34m(minpos, *args)\u001b[0m\n\u001b[1;32m    809\u001b[0m \u001b[38;5;129m@staticmethod\u001b[39m\n\u001b[1;32m    810\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mfrom_extents\u001b[39m(\u001b[38;5;241m*\u001b[39margs, minpos\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[1;32m    811\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    812\u001b[0m \u001b[38;5;124;03m    Create a new Bbox from *left*, *bottom*, *right* and *top*.\u001b[39;00m\n\u001b[1;32m    813\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    824\u001b[0m \u001b[38;5;124;03m       scales where negative bounds result in floating point errors.\u001b[39;00m\n\u001b[1;32m    825\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 826\u001b[0m     bbox \u001b[38;5;241m=\u001b[39m Bbox(\u001b[43mnp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mreshape\u001b[49m\u001b[43m(\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m2\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m2\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[1;32m    827\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m minpos \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    828\u001b[0m         bbox\u001b[38;5;241m.\u001b[39m_minpos[:] \u001b[38;5;241m=\u001b[39m minpos\n",
      "File \u001b[0;32m/scratch/midway3/chaodai/miniconda3/envs/sos/lib/python3.9/site-packages/numpy/core/fromnumeric.py:285\u001b[0m, in \u001b[0;36mreshape\u001b[0;34m(a, newshape, order)\u001b[0m\n\u001b[1;32m    200\u001b[0m \u001b[38;5;129m@array_function_dispatch\u001b[39m(_reshape_dispatcher)\n\u001b[1;32m    201\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mreshape\u001b[39m(a, newshape, order\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mC\u001b[39m\u001b[38;5;124m'\u001b[39m):\n\u001b[1;32m    202\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    203\u001b[0m \u001b[38;5;124;03m    Gives a new shape to an array without changing its data.\u001b[39;00m\n\u001b[1;32m    204\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    283\u001b[0m \u001b[38;5;124;03m           [5, 6]])\u001b[39;00m\n\u001b[1;32m    284\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 285\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_wrapfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43ma\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mreshape\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnewshape\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43morder\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43morder\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/scratch/midway3/chaodai/miniconda3/envs/sos/lib/python3.9/site-packages/numpy/core/fromnumeric.py:56\u001b[0m, in \u001b[0;36m_wrapfunc\u001b[0;34m(obj, method, *args, **kwds)\u001b[0m\n\u001b[1;32m     54\u001b[0m bound \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(obj, method, \u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[1;32m     55\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m bound \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m---> 56\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_wrapit\u001b[49m\u001b[43m(\u001b[49m\u001b[43mobj\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     58\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m     59\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m bound(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwds)\n",
      "File \u001b[0;32m/scratch/midway3/chaodai/miniconda3/envs/sos/lib/python3.9/site-packages/numpy/core/fromnumeric.py:45\u001b[0m, in \u001b[0;36m_wrapit\u001b[0;34m(obj, method, *args, **kwds)\u001b[0m\n\u001b[1;32m     43\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mAttributeError\u001b[39;00m:\n\u001b[1;32m     44\u001b[0m     wrap \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m---> 45\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(\u001b[43masarray\u001b[49m\u001b[43m(\u001b[49m\u001b[43mobj\u001b[49m\u001b[43m)\u001b[49m, method)(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwds)\n\u001b[1;32m     46\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m wrap:\n\u001b[1;32m     47\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(result, mu\u001b[38;5;241m.\u001b[39mndarray):\n",
      "\u001b[0;31mValueError\u001b[0m: setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (4,) + inhomogeneous part."
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAQ8AAAFXCAYAAAC88sukAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8o6BhiAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAx3ElEQVR4nO3de3gTBbo/8G/apunFJL2lJSlpm+6CUGQLKLDormeFssiDdbuI1ssiovisWgUX1q3sEZGDWuse2QXlchbXy0HR7lmBPbo/0T6AogdULsVluVloLKH0EkqY9JqGZn5/lISmTdpcZjIzmffzPPPwNJ1kXujbl5lJ5jsKlmVZEEJIkGKELoAQIk00PAghIaHhQQgJCQ0PQkhIaHgQQkJCw4MQEhIaHoSQkMQJXcBALpcL58+fh1qthkKhELocEiVYlkVbWxsMBgNiYvr+zxR7r0WiZl/bCObJomKxWFgAtNDCy2KxWCTXa5Gouf82AiW6PQ+1Wg0AsFgs0Gg0vG/PbDZj5cqVWLVqFUwmE+/bI8Kw2+0wGo2e/gIi32vBikTNvrYRKNEND/eumEajicgPVK1WQ6lUQq1Wi7KB+ORkLsC2/2OkTpsNpTZD6HIiov+ufqR7LVSRqDmUQyA6YSpTTuYCTv3HfFi2VODUf8yHk7kgdElEYmh4yJB7cDhazgEAHC3naICQoAU9PPbu3Yvi4mIYDAYoFArs2LFj0DonTpzA7bffDq1Wi+TkZEyePBlnz57lol4yDKfNik7zcb+L/eg+nFxRCkeLBXD19j3J1QtHiwUnV5TCfnSfZ12nzSro30WKvSbFmkMV9DmPjo4OFBYW4sEHH8TcuXMHff/MmTP4yU9+goceegirVq2CRqPBsWPHkJCQwEnBXMvNzcXbb78NlUoldCmcsO6uQuO2DcE/0eVCT2sjal9a5HlIP/cxGO54nMPqgiPFXpNizaFSsGzoeR4KhQLbt29HSUmJ57G7774bSqUSW7ZsCek17XY7tFotGIYR9UkssXLarHBe8r/HcLn9Euo3r0CPrRlwua5+IyYG8alZyH14NeKuSQEAKFN0UKbqeK44MOH2mq++4rvXpFBzOK/H6TkPl8uFf/zjHxg9ejRmzZqFzMxMTJ061eeum5vD4YDdbvdaIqmxsRHPP/88GhsbI7pdvihTdUgyFfhdNONvxJjVVVBlGoGY2L4nxcRClWnEmNVV0Iy/0bOuWAaHL6H0GgBBe02KNQ+F0+HR0tKC9vZ2vPTSS7j11lvx6aef4pe//CXmzp2Lzz//3OdzKioqoNVqPYvRaOSypGF1dXXh22+/RVdXF2/bqDbbULLtOCa8eRgl246j2mzjbVuBUGozcO2zW6DKHAkAUGWOxLXPbpHU27Wh9BoAGI1GwXpNijUPhdPPebiu7Ab/4he/wG9+8xsAwIQJE7Bv3z5s2rQJ//Zv/zboOcuXL8fSpUs9X7s/tBItqs02LNlVBwX6PspXe7ELS3bVYe2MfMw0pQpWl3uASPVzHqH0GuD94apI95oUax4Kp3seGRkZiIuLQ0FBgdfjY8eO9Xs2WaVSeT7wIvYP64RifU2jZ3Dgyp8KABtqhD9MUmozkHnrfMkNDiC0XgMgaK9JseahcDo84uPjMXnyZJw6dcrr8e+++w65ublcbkoyvme6MfCMNAvAzHQLUU7UkGKvSbHmoQR92NLe3o7Tp097vjabzThy5AjS0tKQk5ODp556CqWlpbj55ptxyy23YOfOnfjwww/x2WefcVk3Z9LT0/HQQw8hPT2dl9fP0yag9mKX1wBRADBppffWXKRJsdekWHPIgr2Sbs+ePT6vyluwYIFnnb/85S/sD3/4QzYhIYEtLCxkd+zYEfDrMwzDAmAZhgm2NFH6tO4iO3bzQbZg80GvP6vNF4UuTfS47DVffcVHr0mt5nBeL6zPefAh0p/zaGtrQ01NDSZOnBjSlYWBqDbbsKGmEWamGyZtAsom6VGUJ9zJUjkS4nMe4RL75zxEd1VtpFmtVrz66quorKzkbXjMNKUK+s4KIXygC+MIISGh4UEICQkND0JISGQ/PFQqFUaNGhU1V9VyzclcQMvOLZT1QQaR/QnT7OxsvPjii0KXIUqe0KCmerRUb5Xc9S+EX7Lf8yC+UdoYGY7s9zzq6upQXl6OyspK5OfnC13OsHkckeAz86Nf2lj/zA9fxJQDQvgj++EhNiEngUWCj7QxX4ROICORQcNDZHTTS5EyabqgNQSTNuaLMoX2OuSAhofIKFPD3+WvNtuwvqYR3zPdyNMmoGyiPuhPuI5ZXXX1nIer90ramPRCgwh/6IRplHGHD9Ve7EJPL+sJHwo2vSwa0sYIv2S/5zFy5EisW7eOt0vy+WTtdMLa6fR6bM2BBp/hQ2sONCBbffWzLLokJXRJyiFfX+ppY4Rfsh8e8fHx0Ov1QpcRkqoT1oASyVgA9XYH5u044XnssYl6PH69YdjnutPGCBlI9sOjubkZVVVVKC0tRVZWltDlBKV0rA7Tc1O8Hlu2uw5n7Y5B4UM5GhVemX71rejh9joIGY7sh0dHRwe++OIL3HbbbUKXEjRfhx5LJ2d7BS67/1w2JRsFGUkCVEmiFZ0wjTIzTalYOyMfo9MSER+rwOi0RKwryqfwIcI52e95RCMKHyKRQHsehJCQyH54pKam4s4770RqKv1PTUgwZH/YkpqairvuukvoMgiRHNnveXR2duLIkSPo7OwUuhRCJEX2w6OpqQkvvPACmpqahC4lalD6mDzIfngQbrlDhCxbKig8KMrR8CCcofQxeQl6eOzduxfFxcUwGAxQKBTYsWOH33UfeeQRKBQK/OlPfwqjRCJGTpsVnebjnsV+dB9OriiFo8XSdwk/4JU+Zj+6D53m43DaAk9Jk2KvSbHmUAX9bktHRwcKCwvx4IMPYu7cuX7X2759O7766isYDMNffCUkpVKJrKwsKJV0rUcwAk48G5A+FkzKmBR7TYo1hyro4TF79mzMnj17yHUaGhrwxBNP4JNPPsGcOXOGXNfhcMDhcHi+ttvtwZYUFqPRiNdeey2i24wGAxPPAk0fCyZljOteA7z7i49ek2LNoeL8cx4ulwvz58/HU089hXHjxg27fkVFBVatWsV1GVGNi6SwcPlKPIt0+liwvQb0/WchJCnW7A/nJ0wrKysRFxeHxYsXB7T+8uXLwTCMZ7FYLFyXNKT6+no8+OCDqK+vj+h2Q8VVUhgfIp0+FmyvAYDFYhGs1wBp1uwPp3sehw4dwtq1a3H48GEoFIqAnqNSqQS9W1tvby/a2trQ29srWA2BcKeGDZUUNiHrGsFzOiKVPhZKrwGARqOBRqPhpabhSLHmoXC65/HFF1+gpaUFOTk5iIuLQ1xcHOrr67Fs2TLk5eVxuSnZqTphxbwdJ1A/IOgHuJoUVnVC2Pu9uLnTx/iMLZRir0mx5qFwuucxf/58FBUVeT02a9YszJ8/HwsXLuRyU7LjTg0bKimsdKx8bnkgxV6TYs1DCXp4tLe34/Tp056vzWYzjhw5grS0NOTk5AwKElYqlRgxYgSuvfba8KuVMXdq2FBJYUIfsnBNir0mxZpDxgZpz549LPr61WtZsGCBz/Vzc3PZP/7xjwG/PsMwLACWYZhgSwtJZ2cne/LkSbazszMi2+PCp3UX2ZIPjrGFbxxiSz44xlabLwpdEi+47DVffcVHr0mt5nBeT8Gy7MBDaEHZ7XZotVowDCPKk0REmnz1ldh7LRI1h/N6sr+2pbW1FW+99RZaW1uFLoUQSZH98GAYBv/4xz/AMIzQpRAiKbIfHoSQ0NDwIKJH4ULiRMODiBqFC4mX7IeHWq3GrFmzoFarhS6FDEDhQuIm+/R0nU6HRYsWCV0GQV/AkPNS30fsfV7i3y9cqP8l/gOv7iWRIfvh4XA40NDQgOzsbEEv0CMBBgyFES5EuCX74dHQ0IDy8nJUVlYiPz9/+CcQ3vQPGOIjXIhwS/bDg4jHwIChSIcLkeDI/oSpHFWbbSjZdhwT3jyMkm3HRREk5Eukw4VIcGjPQ2bcSWTuq3HdSWRrZ+RHPMowEJEKFyLBk/3wUCgUSExMDCrZSUrcCWRuYk8i88UdLkTERfbDw2Qy4b//+7+FLoM3VSes2FDTOOQ6/ZPIHr9eurcCIJEl++ER7dwJZG6UREa4IvsTphaLBb/5zW9ElUrNJV2SEgUZSZ5l6eRsz6EKEN1JZIRfsh8eTqcT586dg9PpHH7lKDDTlIq1M/IxOi0R8bEKjE5LxLqifBTlie9kKRE3OmyRoZmmVFG+s0KkRfZ7HoSQ0NDwIISERPbDIysrC7/73e+QlZUldCmESIrsh0dycjImT56M5ORkoUshQaB0MeHJfnjYbDZs374dNps4r+8gg1G6mDjQ8LDZsHXrVhoeEkHpYuIR9PDYu3cviouLYTAYoFAosGPHDs/3nE4nysvLMX78eCQnJ8NgMOD+++/H+fPnuayZyISn1/QjoFAosHXtSzi5ohSOFgsuX76MdTUtKP2wFpPXfoLs7GzcUzwb9cePiqNmGfx+BD08Ojo6UFhYiPXr1w/6XmdnJw4fPowVK1bg8OHD2LZtG06dOoXbb7+dk2KJvLh7bfX8YgBA4wfr0dPaCLhc6L7swklbNxZdl4F3Zpvw8k0GHN3/GW4vniOKmuXw+xH0h8Rmz56N2bNn+/yeVqtFdXW112OvvfYapkyZgrNnzyInJye0KoksuXvNabNi0R9eh/6OMsSfqUaPrRnXxAMbpl/ppyvpYq8+8Aym37tI0F6T0+8H758wZRgGCoUCKSkpPr/vcDjgcDg8X9vtdr5L8pKcnIwf//jHsnm3pdpsw/qaRnzPdCNPm4CyiXrRf9rUnS6WlDsGYx5Y5Ddd7PMDR4bsNcC7vyLda74M9/sBiK9mN15PmHZ3d6O8vBz33HOP35voVlRUQKvVehaj0chnSYNkZWVh2bJlsvichzsIqPZiF3p6WU8QkFiTxHzxly7Wq7pm2F4DAKPRKFivDRTI7wcgrpr7423Pw+l04q677gLLsti4caPf9ZYvX46lS5d6vrbb7RH9B3I6nbDb7dBoNFAqo+eq0oEhQMDQQUDZahV0SUpJXFk7MF0MSVrcdccdw/Ya0HcVdf87zgv1yxjo7wcgnpoH4mV4uP9h6uvrsXv37iGnqkqlEvSWBxaLJSrT0wMJAQKuBgHN23ECj03USyYMyJ0uFkyvAYBGoxl2Hb5JsWZfOB8e7n+Y2tpa7NmzB+np6VxvggRgYAgQMHQQ0CvT8yWx19GfFHtNijX7E/TwaG9vx+nTpz1fm81mHDlyBGlpadDr9Zg3bx4OHz6Mjz76CL29vWhqagIApKWlIT4+nrvKyZB8HYIsnZztFX7cPwioICNJgCqHJsVek2LNIWODtGfPHhZ9Pee1LFiwgDWbzT6/B4Dds2dPQK/PMAwLgGUYJtjSQnLmzBl23rx57JkzZyKyPaF9WneRLfngGFv4xiG25INjbLX5otAl+cVlr/nqKz56TWo1h/N6Qe95/OxnPwPLsn6/P9T3iPCkFAQkxV6TYs2hkn2SWF5eHrZu3YrY2FihSyFEUmQ/PGJiYhATI/vrAwkJmux/a86fP4+VK1dK9uIkQoQi++HR3d2N48ePo7u7W+hSCJEU2Q8PKaM0LSIkGh4SRWlaRGg0PCSI0rSIGMj+3ZaMjAw88sgjyMjIELqUQZw2K5yXrF6PXW6/hPrNK9BjawZcrr4HXb1wtFhwckUpch9ejbhrUqBM0XkuZSeED7IfHhqNBjNmzBC6DJ+su6vQuG1DYCu7XOhpbUTtS4sAAPq5j8Fwx+M8VkfkTvbDw26348CBA5g8ebLorlzUTS9FyqTpXo/53PMAPGla/fc8COGT7IfHhQsXsGnTJphMJl6GRzjJXcpU34ceY1ZX+U3TUmrFd/hFohOdMOURX8ld/tK0aHCQSJL9ngdfrJ1Ov8ldG2oaw744bWCaFg0OEmk0PHhSdcKKertj0OMsADPDzadZ3WlahAhB9octCQkJKCgoQEJCAqevWzpWh1yNCooBjysAmLTcbosQIch+eBgMBqxatQoGA7fZnbokJZZOzvYcqgBXk7vKJuk53RYhQpD98HC5XHA6nXD1f9uTIzNNqVg7Ix+j0xIRH6vA6LRErCvKR1GeNMJ4CBmK7M95fP/997ymp0spuYuQYMh+z4MQEhoaHoSQkNDwIISEhIYHoVAhEhLZnzA1Go3YtGmT6C6KixRPNkhTPVqqt9LH3EnAZL/noVQqkZ6eHlU3uQ4UhQqRcAQ9PPbu3Yvi4mIYDAYoFArs2LHD6/ssy+LZZ5+FXq9HYmIiioqKUFtby1W9nGtubsYrr7yC5uZmoUvhnNNmRaf5uM/FfnQfTq4ohaPF0ndlLuAVKmQ/ug+d5uNw2qxDb4RHUuw1KdYcqqCHR0dHBwoLC7F+/Xqf33/55Zexbt06bNq0CV9//TWSk5Mxa9Ys0aaTd3R04KuvvkJHR4fQpXDOursKJ56Z53OpfWkRelobvTNBAK9QoRPPzIN1d5UwxUOavSbFmkMWzn0uAbDbt2/3fO1yudgRI0awf/jDHzyPXbp0iVWpVOx7770X0GvSvWq503Oxhe2oO+ZzYf75f+w/n5jOHvzVOPbgvWOvLr8ax/7zieks88//YzvqjrE9F1uE/muwLBt+r0XqXrVSqzmi96oditlsRlNTE4qKijyPabVaTJ06Ffv378fdd9896DkOhwMOx9WrT+12O5clyZq/MCE3KYcKhdJrgHd/RbrXpFjzUDgdHk1NTQCArKwsr8ezsrI83xuooqICq1at4rIM0QonVYwP7kwQ97stUhkcQGi9BvS9uyYUKdY8FMHfbVm+fDkYhvEsFoslottPTU3Fvffei9RUfn+J+UoVC5d7gBjnL5fM4AiHxWIRrNdCJdaaOR0eI0aMAIBB71w0Nzd7vjeQSqWCRqPxWiIpNTUVv/zlL3kfHusOnfeZKrbukPD3yHWHCklpcITSawAE7TUp1jwUToeHyWTCiBEjsGvXLs9jdrsdX3/9NaZNm8blpjjT0dGBAwcO8P5uy/dMt2dwuLFXHifBk2KvSbHmoQR9zqO9vR2nT5/2fG02m3HkyBGkpaUhJycHTz75JJ5//nmMGjUKJpMJK1asgMFgQElJCZd1c6a5uRkvv/wyb5fku+VpE1B3yXuAKK48TnyTYq9JseaQBfv2zJ49e1j0/afptSxYsIBl2b63o1asWMFmZWWxKpWKnTFjBnvq1KmAXz9a36r9tO4iO3bzQbZg80GvP6vNF3ndrpRx2WuReqtWajWH83oKlmUH7k0Lym63Q6vVgmGYiBzf1dXV8RoG1F+12YYNNY0wM90waRNQNklPqWIR4quvIt1rwYpEzeG8nuwvjIskShUj0UTwt2qFplQqMXLkSFleGEdIOGS/52E0GvHHP/5R6DIIkRzZ73kQQkIj++FhNptx//33w2w287YNSuoi0Uj2w4NlWXR1dYGvN53cgTuWLRUUtEOiiuyHB58oqYtEM9mfMOWK02aF89LV1K3L7ZdQv3kFemzNVwN3+iV15T68GokjRw15yTwhYkbDgyPW3VVo3LZh+BX7JXXp5z4Gwx2P818cITyQ/fDIzs5GZWUlsrOzw3od3fRSpEya7vna554HAMTEID41y7PnQYhUyX54qFSqIT+WHmiAj6/ULikndREyHNmfMLVarXj99ddhtQ5OCQ83wMcdtKPKHAkANDhIVJH98Ghra8Mnn3yCtra2Qd9bX9PoM8BnQ01jwK8vt6QuIh+yP2wZir8AH3OQAT7upC5Coons9zyGkqdNgGLAYwoAJgrwIYSGx1DKJuo9hyq48icLoGySXriiCBEJ2Q8PrVaLOXPmQKvVDvreTFMq1s7Ix+i0RMTHKjA6LRHrivIpwIcQ0DkPpKen44EHHvD7fQrwIcQ32e95dHV14dSpU+jq6hK6FEIkRfbDo7GxEc888wwaGwN/+5UQQsODEBIiGh6EkJDQ8CDDoiQ04ovsh0dsbCzUajViY2OFLkWUKAmN+CP74ZGbm4s33ngDubm5QpciOpSERobC+fDo7e3FihUrYDKZkJiYiB/84AdYvXo1bxmhJDROmxWd5uN+F/vRfTi5ohSOFktfnADglYRmP7rPs67TNviK5EiQYq9JsWZ/OP+QWGVlJTZu3Ii3334b48aNw8GDB7Fw4UJotVosXryY682FzWKxoLKyEuXl5TAajUKXEzEBJ58N1C8JzU2oRDSp9RogzZr94Xx47Nu3D7/4xS8wZ84cAEBeXh7ee+89fPPNNz7XdzgccDgcnq/tdjvXJQ3J6XSiubkZTqczotsV2sDks4ECSUKLuyYFAKBMESaHNdheA7z7K9K9BkizZn84P2y58cYbsWvXLnz33XcAgG+//RZffvklZs+e7XP9iooKaLVazyKn//3DVW22oWTbcUx48zBKth0POKQI6Es+SzIV+F0042/EmNVVUGUagZgrJ5NjYqHKNGLM6ipoxt/oWVeoEOdgew3ou0OgkL0mxZr9YjnW29vLlpeXswqFgo2Li2MVCgX74osv+l2/u7ubZRjGs1gsFhYAyzAM16X5dObMGXbevHnsmTNnIrI9rnxad5Edu/kgW7D5oNefn9Zd5HQ7PZes7NGlt7IH7x3LHl16K9tzycrp64cjmF5jGIYFwFosFsF6TYw1u7cRyutxftjy17/+Fe+++y62bt2KcePG4ciRI3jyySdhMBiwYMGCQeurVCqoVCquy4g61k4nrJ1XD63WHGjwmXK25kADstUq6JKU0CWFf/NudxKabf/HSJ02W1RJaMH2GgBoNBpoNJoIV3qVFGv2R8Gy3J7mNRqNePrpp1FWVuZ57Pnnn8c777yDkydPDvt8u90OrVYLhmEi8g/W2dmJ7777DqNHj0ZSUhLv2wvVa4fOBxV/+NhEPR6/3sBjRcILptd89VWke02MNYfzepzveXR2diImxvtUSmxsLFz9T7qJSFJSEiZMmCB0GcMqHavD9NwUz9fLdtfhrN3hFZOoAJCjUeGV6fmc7HWIndR6DZBmzf5wPjyKi4vxwgsvICcnB+PGjUNNTQ3WrFmDBx98kOtNccJms6G6uhozZ85Eaqp4czsGHoYsnZyNJbvqPIcu7j+XTclGQYZ496C4JLVeA6RZs1+cnHXpx263s0uWLGFzcnLYhIQENj8/n/33f/931uFwBPT8cE7ghEKqJ0xZtu+kackHx9jCNw6xJR8cY6vN3J4sFbtges1XX0W618RYczivx/k5j3BF+ji0rq4O5eXlqKysHPLmT0TaxHLOIxhiP+ch+2tbCCGhoeFBCAmJ7IdHcnIyfvrTnyI5OVnoUgiRFNkPj6ysLCxevBhZWVlCl8IpCvAhfJP98Ojp6UFjYyN6enqELoUzFOBDIkH2w+PcuXNYvHgxzp07J3QpnKAAHxIpsr/pk9Q4bVY4L/kO3/F5GX2/AJ+Bl9ELdTUsiQ40PCQmpBAfEQX4kOhBw0NihgrxkUKAD4keNDwkRpk69OHGmNVVV895uHqvBPiMxLXPbhHV5fRE+mR/wjQ/Px//8z//w+tH08NJ/AqWO39DlTkSAGhwEN7QngfPqs02r6tfay92YcmuOqydkY+ZJn6u4hVzgA+JHrLf82hoaMDvf/97NDQ08PL662safSZ+BRPsEwqlNgOZt86nwUF4I/vh4XA4UFtb65XgzqXvmW4MvGyZBWBmunnZHiGRIvvhwbc8bQIUAx5TADBpE4QohxDO0PDgWdlEvedQBbia+FU2SS9cUYRwgIYHz2aaUrF2Rj5GpyUiPlaB0WmJWFeUj6I88UYeEhII2b/botPp8MQTT0Cn4+9DUzNNqby9s0KIUGQ/PNRqNW6++WahyyBEcmR/2MIwDHbu3AmGYYQuhRBJkf3waG1txV/+8he0trYKXQohkiL74SFGlAJGpICGh8hQChiRChoeIkIpYERKeBkeDQ0N+NWvfoX09HQkJiZi/PjxOHjwIB+bCltiYiIKCwuRmJgYke05bVZ0mo8PWuxH9+HkilI4Wix9l9IDXilg9qP7POs6bb6TxORISr3mJsWafeH8rVqbzYabbroJt9xyCz7++GPodDrU1taK9j6wer0ezzzzTMS2F3QSGKWA+SW1XgOkWbM/nA+PyspKGI1GvPnmm57HTCYT15vhTG9vLxwOB1QqFWJjY3nfnr8kMEoBC57Ueg2QZs3+cH7Y8r//+7+44YYbcOeddyIzMxMTJ07E5s2b/a7vcDhgt9u9lkiqr6/HggULUF9fH5HtKVN1SDIVIMlUgP+DHvfWADd+1o37a1PQ/PCfoco0AjFXhlhMLFSZRoxZXQXN+Bs9z6Pg4j7B9hoAQXsNkGbN/nA+POrq6rBx40aMGjUKn3zyCR599FEsXrwYb7/9ts/1KyoqoNVqPYvRaOS6JFFyhwTVXuxCTy+L2otdWPz1JTQ98BqlgAUo2F4DAKPRKGivSbFmfxQsyw6MmwhLfHw8brjhBuzbt8/z2OLFi3HgwAHs379/0PoOh8MrS8Nut8NoNEbszuV1dXUoLy9HZWUlr1GEAxX/7RjqLnlnfSgA5KckYNvMLEoBC0Awvea+G7zFYvG643wke02MNbu3Ecrrcb7nodfrUVBQ4PXY2LFjcfbsWZ/rq1QqaDQar0UO/IUEfc90UwpYgILtNQCC95oUa/aH8+Fx00034dSpU16Pfffdd8jNzeV6U5LmLyQoj0KCAibFXpNizX6xHPvmm2/YuLg49oUXXmBra2vZd999l01KSmLfeeedgJ7PMAwLgGUYhuvSfHI6neylS5dYp9MZke25fVp3kR27+SBbsPmg15/V5osRrUPKguk1X30V6V4TY83hvB7nw4NlWfbDDz9kr7vuOlalUrFjxoxh//znPwf8XCF+oEL5tO4iW/LBMbbwjUNsyQfHaHCEINBeE8vwYFlx1RzO63F+wjRc4ZzACUVTUxPeeustPPDAAxgxYgTv2yPC8NVXke61YEWiZlGdMJWazs5OHDp0CJ2dnUKXQoikyH54EEJCQ8ODEBISGh6EkJDIZnj4S+dKS0vD/fffj7S0NIEqI0SaZJGe7gnZaapHS/VWr+tFUlJSUFxcLHCFhEhP1O95DJfO1d7ejv3796O9vV3IMgmRnKgdHk6bddh0rs76U2hpacGaNWvQ0tIibMGESEzUDg/r7irUvrQIPa2N3uE6gCedq+H9V4QpjpAoELXDQze9FKOefh3x6XogZsBfMyYG8el6ZN+9TJjiCIkCUTs8lKk6aMbfiDGrq/ymcyXlXutZf19LD0q2HceENw+jZNtxVJttAlVOiDRE7fBwU2ozcO2zW/ymc8XHxyNh/E/w4r86vFK9luyqowFCyBBk8Vate4C407kuKbWwXrhyLUtCGi6MmQGF3eEJ52HRl62x5kADstUq6JKU0CUpBaqeEHGSxfAA4EnnAoCqQ+exoaZxyPVZAPV2B+btOIHHJurx+PWGCFRJiHTIZnj0VzpWh+m5KQD6bsDz1OffoyfJ+xOmCgA5GhVemZ5Pex2E+CDL4dH/MCTBHousU5/BMnEuFLh6yMICWDYlGwUZSQJWSoh4Rf0J00BoWr7D769Lxui0RMTHKjA6LRHrivJRlCe9u3gREimy3PPw5cbMePzqx5G79QIhUkd7HoSQkMh+zyM7Oxtr1qxBZmam0KUQIimyHx4qlUpUt/AjRCqi4rDFX9BPIKxWKzZu3Air1cpDZYREL8kPD3deh2VLhVdOR6Da2tqwe/dutLW18VQhIdFJ0sNjuKAfQgh/eB8eL730EhQKBZ588smwXsdps6LTfNyzDBf0Yz+6D53m43Da6HBELrjqtUiSYs1uvJ4wPXDgAP7rv/4LP/rRj8J+LevuKjRu2zD8ileCfmpfWgQA0M99DIY7Hg97+0TcuOy1SJFizf3xtufR3t6O++67D5s3b0Zqavif1NRNL8XY5//mWYYL+hn19OsY+/zfoJteOuTrarValJSUQKvVhl0jEQbXvRYJUqx5IN6GR1lZGebMmYOioqIh13M4HLDb7V6LL8pUHZJMBZ5luKAfzfgbkWQqgDJVN+T209PTcd999yE9PT2kvycRXqC9BiCgXosEKdY8EC+HLe+//z4OHz6MAwcODLtuRUUFVq1aFdJ23Dkd7tsqDAz6cas227C+phHfM93I0yagbKIeM019076rqwt1dXXIz89HYmJiSHUQ4QTTawBE8ZkeKdbsC+d7HhaLBUuWLMG7776LhISEYddfvnw5GIbxLBaLJajtuQeIcf5yv4Njya46vylhjY2NeO6559DYOHS+BxGfYHvN/ZxQe40LUqzZH873PA4dOoSWlhZMmjTJ81hvby/27t2L1157DQ6HA7GxsZ7vqVQqqFSqsLbZP+hnoHWHznsusQeuXnK/7tB5z94HkaZgew0ANBoNNBpNpEv1kGLN/nA+PGbMmIGjR496PbZw4UKMGTMG5eXlg/5h+PY90+0ZHG7slceJtImt1wIhxZr94Xx4qNVqXHfddV6PJScnIz09fdDjkZCnTUDdJe8BorjyOJE2sfVaIKRYsz+S/oRpIBZfb/AcqgBXU8KW3NCXSRobG4u0tDRJTXxCxEDBsuzAvXpB2e12aLVaMAzD2XFetdmGDTWNMDPdMGkTUDZJTylhMuOrr/joNS5FouZwXk8Wl+TPNKXSyVFCOBb1hy3Dqa+vx69//WvU19cLXQohkiL74dHb24uLFy+it7dX6FIIkRTZDw9CSGhkNTzCSRwjhHiTzfAIN3GMEOJNFsNjqMQxvV6P5557Dnq9XuAqCZGWqB8eTuYCTq6813fi2Mp7EdfTgXHjxtEVtYQEKeqHh23/x+ixngNcLu9vuFzosZ7DuV0f4N1330Vra6swBRIiUVE/PFKnzUa8bqTvxDHdSPT+8Abs2LEDDMMIUyAhEhX1w0OpzcCYVVt9J46t2oqYa/x/8rTabEPJtuOY8OZhlGw77skAIYTIYHgAVwODVJkjAcBv4lh/w4UIESJ3sri2Bbg6QGz7P0bqtNk+B4e10wlrpxMAsOZAg88QoTUHGjAh6xrokpSRKp0QUZLN8AB8J46p1WpMnz4darUaVSes2FDjP46QBVBvd6DqhBWPX2/guVpCxE1Ww8MXnU6HRx99FABQmuzE9NwUAMCy3XU4a3cMChHK0ahQOnboRHZC5EAW5zyG4nA4YLFY4HA4oEtSoiAjCQUZSVg6OdtniNCyKdl0yEIIaHigoaEBS5cuRUNDg9fjM02pWDsjH6PTEhEfq8DotESsK8qnECFCrpD9YctQKESIEP9kv+dBCAkNDQ9CSEhkPzwUCgXi4uKgUCiGX5kQ4iH7cx4mkwnvvfee0GUQIjmy2/OgNDFCuCGr4eErTezcuXP43e9+h3PnzgldHiGSwvnwqKiowOTJk6FWq5GZmYmSkhKcOnWK680EzV+aWPfFZpjNZvT09AhcIQmWWHttKFKs2R/Oh8fnn3+OsrIyfPXVV6iurobT6cTPf/5zdHR0cL2pgA2VJtb2598iCZcFq42EToy9Nhwp1uwP5ydMd+7c6fX1W2+9hczMTBw6dAg333wz15sLiCdNbCCXCy5bM8bG0DstUiTGXhuOFGv2h/d3W9wJXWlpaT6/73A44HA4PF/b7XbOa0idNhvNO7egp/W8dxxhTAxitDqcaNZyvk0SecP1GuDdX3z0WrCkWLMbrze6drlcuP3223Hp0iV8+eWXPtd57rnnsGrVqkGPc33zYa9zHq7eK2liIzFy2SacqD+P8ePHY7/VifU1jfie6UaeNgFlE/X08XSJGK7X3Dd09kWoG12LoeZwbnTN67stZWVl+Ne//oX333/f7zrLly8HwzCexWKx8FKLvzSxFEMupk2bhv1WJyWHSVggvQYAFouF914LlBRr7o+3w5bHH38cH330Efbu3YuRI0f6XU+lUkGlUvFVhpf+aWKuCTNR60xCm7kZNTU12Naj95sclq1WQZekpEvxRSrQXgMAjUYjyF7GQFKseSDOhwfLsnjiiSewfft2fPbZZzCZTFxvIizuNLHXDp3HhpoTVx7VAT7ecXEnh83bcQKPTdRTepjIiL3XfJFizf5wPjzKysqwdetW/P3vf4darUZTUxMAQKvViurGSqVjdZiem4JzDeewbu06dM94CI1dLp/JYa9Mz6e9DhGSSq/1J8Wa/eH8nMfGjRvBMAx+9rOfQa/Xe5aqqiquNxUWd2rYD9VxSGxrxgM/SPSbHFaQkUTDQ4Sk0mv9SbFmf3g5bJGiGzPjsXZGFjbUNMLMdMOkTUDZJD0lh4mYFHtNijX7I/urapOSknD99dcjKSkJM0dQchghgZL98BgxYgSefvppocsgRHJkdVWtL5cvXwbDMLh8ma5vISQYsh8eZ8+exaJFi3D27FmhSyFEUmQ3PCgMiBBuyOqch+f6lqZ6tFRvxbXPbhG6JEIkSzZ7Hv7CgFztdO0KIaGQxfAYdEUtcCUM6BzaXi+nMCBCQiCL4WHb/zEcTfVXB4ebqxeu1vN44f5i5ObmClMcIRIli+GROm02VCNygZhY72/ExEI1IhdZN9+O2NhY308mhPgki+HhleXhHiBXwoBSfv2fqHx1ExobG0N67WqzDSXbjmPCm4dRsu045X8Q2ZDF8AD8hwH1xCXi22+/RVdXV9CvWW22UYAQkS1ZvVXbPwwoddpsKLUZQGtomZDWTifWHGjwGSC0oaaRrpEhUU9WwwO4GgYUrqoTVtTbHYMeZwGYme6wX58QsZPNYQvXSsfqkKtRYeBNGxQATNoEIUoiJKJkPzzS09Px0EMPIT09Pajn6ZKUWDo522eAUNkkPcdVEiI+sh8eWq0Wt956q9+I+6HMNKVi7Yx8jE5LRHysAqPTErGuKJ8ChIgsyO6cx0BtbW2oqanBxIkToVarg37+TBMFCBF5kv2eh9Vqxauvvgqr1Sp0KYRIiuyHByEkNDQ8CCEhoeFBCAmJ7IeHSqXCqFGjPLe8pKQxQgIj+3dbsrOz8eKLLwLwnTSm1GYIXCEh4iT7PQ83f0ljtAdCiG+8DY/169cjLy8PCQkJmDp1Kr755hu+NhWWuro6PHrnbTi6/A44WiwDksYsOLnyXhogIieVXutPijUPxMvwqKqqwtKlS7Fy5UocPnwYhYWFmDVrFlpaWvjYXNhuiWsGy1gBl8v7Gy4XeqznYNv/sTCFkWFJrdcAadbsCy/DY82aNXj44YexcOFCFBQUYNOmTUhKSsIbb7wxaF2HwwG73e61RNqey1lQaHVAzIB/jpgYxOtGInXa7IjXRAITTK8BELzXAGnW7Avnw6OnpweHDh1CUVHR1Y3ExKCoqAj79+8ftH5FRQW0Wq1nMRqNXJc0rAtIhOaRNVBlGgckjRkxZtVWKLUZlBgmQsH2GgAYjUZBe02KNfvD+fC4cOECent7kZWV5fV4VlYWmpqaBq2/fPlyMAzjWSwWC9clBSTmmlSfSWPuwUGJYeITbK8BgMViEbTXpFizP4K/VatSqTyfsRDCyJEjsW7dOqSnp0MZH49rn90Cy96P0DVuBmqdScCFTr+JYWsONCBb3Ve7LkkJXZJSoL8FCZRGo4FGoxG6jKCItWbOh0dGRgZiY2PR3Nzs9XhzczNGjBjB9ebCFh8fD73+av6GUpuB/2f4OTbsbgTg/wQWC6De7sC8HScAAI9N1OPx6w08V0v6k1qvAdKs2R/OD1vi4+Nx/fXXY9euXZ7HXC4Xdu3ahWnTpnG9ubA1Nzdj3bp1Xj/M0rE6/K1krGfxlxiWq1F51ikdq4to3UR6vQZIs2Z/eDlsWbp0KRYsWIAbbrgBU6ZMwZ/+9Cd0dHRg4cKFfGwuLB0dHfjiiy9w2223eR4beAiydHI2luyq8xy6uP9cNiUbBRlJkS6Z9COlXnOTYs2+8DI8SktLYbVa8eyzz6KpqQkTJkzAzp07B50kkgp3YtiGmkaYmW6YtAkom6SnxDARkGKvSbFmXxQsy7LDrxY5drsdWq0WDMNE5CRRXV0dysvLUVlZifz8fN63R4Thq68i3WvBikTN4bye4O+2DOSeZZH6MExbWxucTifa2tpE9QEcwi33z7b//5WR7rVgRaJmX9sIlOiGR1tbGwBE/MMwf//73yO6PSKMtrY2T9h1a2srgMj3WrBaW1t5r7n/NgIluuFhMBhgsVigVquhUAx8j4N7drsdRqMRFotFlLuukRat/x4sy6KtrQ0Gw9W309PS0gAAZ8+eDSk9n28MwyAnJ8dTJ8B9zb62ESjRDY+YmBiMHDky4tsV6wdxhBKN/x4Df9lirlzLpNVqRf13jel3zRVfNccMvK4rkOdwtnVCiKzQ8CCEhET2w0OlUmHlypWCXl8jJnL69xD739VXfVzXHM7rie5zHoQQaZD9ngchJDQ0PAghIaHhQQgJCQ0PQkhIaHgQQkIi6+ERDffO4Mpzzz0HhULhtYwZM0bosngl5p//3r17UVxcDIPBAIVCgR07dgDgtmZ/2wiUbIdHtNw7g0vjxo1DY2OjZ/nyyy+FLok3Yv/5d3R0oLCwEOvXr/c8xnXNvrYRFFampkyZwpaVlXm+7u3tZQ0GA1tRUSFgVcJZuXIlW1hYKHQZESOlnz8Advv27bzW7N5GMGS55xHKvTPkoLa2FgaDAfn5+bjvvvtw9uxZoUvihRR//k6nU3Q1y3J4hHLvjGg3depUvPXWW9i5cyc2btwIs9mMn/70p558lWgixZ9/W1ub6GoW3SX5RBizZ1+9peaPfvQjTJ06Fbm5ufjrX/+Khx56SMDKiFjJcs8jmu6dwZeUlBSMHj0ap0+fFroUzknx569Wq0VXsyyHRzTdO4Mv7e3tOHPmjNcNsaKFFH/+SqVSdDXL9rAlWu6dwZXf/va3KC4uRm5uLs6fP4+VK1ciNjYW99xzj9Cl8ULsP//29navvT6z2Yy5c+di5cqVnNXsaxtHjhxBWloacnJyhn+BsN/jkbBXX32VzcnJYePj49kpU6awX331ldAlCaa0tJTV6/VsfHw8m52dzZaWlrKnT58Wuixeifnnv2fPHhZ99xbzWqZMmcJZzf62sWDBgoCeT3kehJCQyPKcByEkfDQ8CCEhoeFBCAkJDQ9CSEhoeBBCQkLDgxASEhoehJCQ0PAghISEhgchJCQ0PAghIaHhQQgJyf8HR3hPNMo6gkUAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 390x400 with 3 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "eargs = {'fmt':'.', 'ms':8,   'lw':1, 'clip_on':False, 'c':style.qtl_color_s['eqtl'], 'label':'eQTL'}\n",
    "sargs = {'fmt':'D', 'ms':3.5, 'lw':1, 'clip_on':False, 'c':style.qtl_color_s['sqtl'], 'label':'sQTL'}\n",
    "plot_enrichment(eqtl_tissue_avg_df, sqtl_tissue_avg_df, mean_df, std_df, eargs, sargs)\n",
    " # plt.savefig('pdfs/Fig4A_torus_enrichment_summary.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_enrichment(df1, df2, varmean_df, varstd_df, args1, args2, legend_title=None):\n",
    "    dl = 1.25\n",
    "    aw1 = 1.5\n",
    "    ds = 0.05\n",
    "    aw2 = 0.4\n",
    "    dr = 0.5\n",
    "    db = 0.75\n",
    "    ah = 3/17*df1.shape[0]\n",
    "    dt = 0.25\n",
    "    aw3 = 0.2\n",
    "    \n",
    "    fw = dl + aw1 + ds + aw2 + aw3 + dr\n",
    "    fh = db + ah + dt\n",
    "    fig = plt.figure(facecolor=(1,1,1), figsize=(fw,fh))\n",
    "    ax1 = fig.add_axes([dl/fw, db/fh, aw1/fw, ah/fh])\n",
    "    ax2 = fig.add_axes([(dl+aw1+ds)/fw, db/fh, aw2/fw, ah/fh], sharey=ax1)\n",
    "    ax3 = fig.add_axes([(dl+aw1+ds+aw2+0.05)/fw, db/fh, aw3/fw, ah/fh], sharey=ax1)\n",
    "\n",
    "    y = np.arange(df1.shape[0])\n",
    "    ax1.set_ylim([y[0]-0.5, y[-1]+0.5])\n",
    "    \n",
    "    qtl.plot.plot_effects([df1, df2], [args1, args2],\n",
    "                 xspace=[1.25,2,0.5], yspace=[0.5,ah,0.25], xlim=[-2,9],\n",
    "                 xlabel='log$\\mathregular{_{2}}$(Fold enrichment)', ylabel=None, ax=ax1)\n",
    "\n",
    "    ax1.set_yticks(y)\n",
    "    ax1.set_yticklabels(df1.index)\n",
    "    ax1.legend(title=legend_title, loc='upper right', ncol=1, labelspacing=0.33, \n",
    "               columnspacing=1, handletextpad=0, handlelength=1.66, borderpad=0.25, framealpha=1)\n",
    "    for ax in [ax2, ax3]:\n",
    "        plt.setp(ax.get_yticklabels(), visible=False);\n",
    "        for line in ax.yaxis.get_ticklines():\n",
    "            line.set_markersize(0) # tick length\n",
    "            line.set_markeredgewidth(0) # tick line width\n",
    "    ax2.spines['right'].set_visible(False)\n",
    "    ax3.spines['left'].set_visible(False)\n",
    "\n",
    "    # plot proportion of variants\n",
    "    varmean_df.plot(kind='barh', xerr=1.96*varstd_df, ax=ax2, legend=False, width=0.66, \n",
    "                 color=[args1['c'], args2['c']])\n",
    "    varmean_df.plot(kind='barh', xerr=1.96*varstd_df, ax=ax3, legend=False, width=0.66, \n",
    "                 color=[args1['c'], args2['c']])\n",
    "\n",
    "    ax2.set_xlim([0,0.12])\n",
    "    ax2.set_xticks([0,0.1]) \n",
    "    ax3.set_xlim([0.5, 0.62])\n",
    "    ax3.set_xticks([0.5,0.6])\n",
    "    ax2.set_xticklabels([0,0.1], ha='center', fontsize=8)\n",
    "    ax3.set_xticklabels([0.5,0.6], ha='left', fontsize=8)\n",
    "    ax2.set_xlabel('Proportion\\n of variants', ha='left')\n",
    "\n",
    "    ax1.xaxis.set_major_locator(ticker.MaxNLocator(integer=True, min_n_ticks=5, nbins=4))\n",
    "    cis_xlim = ax1.get_xlim()\n",
    "    ax1.invert_yaxis()\n",
    "    ax2.invert_yaxis()\n",
    "    ax3.invert_yaxis()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Below use R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(tidyverse)\n",
    "library(data.table)\n",
    "library(bedtoolsr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "setwd('/project/yangili1/cdai/splice-pub/analysis')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissues = c('Liver')\n",
    "sgene_files = glue::glue('../code/results/qtl/noisy/GTEx/{tissues}/separateNoise/cis_100000/perm/chr{1:22}.addQval.txt.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr1.addQval.txt.gz'</li><li>'../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr2.addQval.txt.gz'</li><li>'../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr3.addQval.txt.gz'</li><li>'../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr4.addQval.txt.gz'</li><li>'../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr5.addQval.txt.gz'</li><li>'../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr6.addQval.txt.gz'</li><li>'../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr7.addQval.txt.gz'</li><li>'../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr8.addQval.txt.gz'</li><li>'../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr9.addQval.txt.gz'</li><li>'../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr10.addQval.txt.gz'</li><li>'../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr11.addQval.txt.gz'</li><li>'../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr12.addQval.txt.gz'</li><li>'../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr13.addQval.txt.gz'</li><li>'../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr14.addQval.txt.gz'</li><li>'../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr15.addQval.txt.gz'</li><li>'../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr16.addQval.txt.gz'</li><li>'../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr17.addQval.txt.gz'</li><li>'../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr18.addQval.txt.gz'</li><li>'../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr19.addQval.txt.gz'</li><li>'../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr20.addQval.txt.gz'</li><li>'../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr21.addQval.txt.gz'</li><li>'../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr22.addQval.txt.gz'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis\\_100000/perm/chr1.addQval.txt.gz'\n",
       "\\item '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis\\_100000/perm/chr2.addQval.txt.gz'\n",
       "\\item '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis\\_100000/perm/chr3.addQval.txt.gz'\n",
       "\\item '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis\\_100000/perm/chr4.addQval.txt.gz'\n",
       "\\item '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis\\_100000/perm/chr5.addQval.txt.gz'\n",
       "\\item '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis\\_100000/perm/chr6.addQval.txt.gz'\n",
       "\\item '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis\\_100000/perm/chr7.addQval.txt.gz'\n",
       "\\item '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis\\_100000/perm/chr8.addQval.txt.gz'\n",
       "\\item '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis\\_100000/perm/chr9.addQval.txt.gz'\n",
       "\\item '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis\\_100000/perm/chr10.addQval.txt.gz'\n",
       "\\item '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis\\_100000/perm/chr11.addQval.txt.gz'\n",
       "\\item '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis\\_100000/perm/chr12.addQval.txt.gz'\n",
       "\\item '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis\\_100000/perm/chr13.addQval.txt.gz'\n",
       "\\item '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis\\_100000/perm/chr14.addQval.txt.gz'\n",
       "\\item '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis\\_100000/perm/chr15.addQval.txt.gz'\n",
       "\\item '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis\\_100000/perm/chr16.addQval.txt.gz'\n",
       "\\item '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis\\_100000/perm/chr17.addQval.txt.gz'\n",
       "\\item '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis\\_100000/perm/chr18.addQval.txt.gz'\n",
       "\\item '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis\\_100000/perm/chr19.addQval.txt.gz'\n",
       "\\item '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis\\_100000/perm/chr20.addQval.txt.gz'\n",
       "\\item '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis\\_100000/perm/chr21.addQval.txt.gz'\n",
       "\\item '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis\\_100000/perm/chr22.addQval.txt.gz'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr1.addQval.txt.gz'\n",
       "2. '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr2.addQval.txt.gz'\n",
       "3. '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr3.addQval.txt.gz'\n",
       "4. '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr4.addQval.txt.gz'\n",
       "5. '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr5.addQval.txt.gz'\n",
       "6. '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr6.addQval.txt.gz'\n",
       "7. '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr7.addQval.txt.gz'\n",
       "8. '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr8.addQval.txt.gz'\n",
       "9. '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr9.addQval.txt.gz'\n",
       "10. '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr10.addQval.txt.gz'\n",
       "11. '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr11.addQval.txt.gz'\n",
       "12. '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr12.addQval.txt.gz'\n",
       "13. '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr13.addQval.txt.gz'\n",
       "14. '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr14.addQval.txt.gz'\n",
       "15. '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr15.addQval.txt.gz'\n",
       "16. '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr16.addQval.txt.gz'\n",
       "17. '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr17.addQval.txt.gz'\n",
       "18. '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr18.addQval.txt.gz'\n",
       "19. '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr19.addQval.txt.gz'\n",
       "20. '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr20.addQval.txt.gz'\n",
       "21. '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr21.addQval.txt.gz'\n",
       "22. '../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr22.addQval.txt.gz'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr1.addQval.txt.gz\n",
       "../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr2.addQval.txt.gz\n",
       "../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr3.addQval.txt.gz\n",
       "../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr4.addQval.txt.gz\n",
       "../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr5.addQval.txt.gz\n",
       "../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr6.addQval.txt.gz\n",
       "../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr7.addQval.txt.gz\n",
       "../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr8.addQval.txt.gz\n",
       "../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr9.addQval.txt.gz\n",
       "../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr10.addQval.txt.gz\n",
       "../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr11.addQval.txt.gz\n",
       "../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr12.addQval.txt.gz\n",
       "../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr13.addQval.txt.gz\n",
       "../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr14.addQval.txt.gz\n",
       "../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr15.addQval.txt.gz\n",
       "../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr16.addQval.txt.gz\n",
       "../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr17.addQval.txt.gz\n",
       "../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr18.addQval.txt.gz\n",
       "../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr19.addQval.txt.gz\n",
       "../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr20.addQval.txt.gz\n",
       "../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr21.addQval.txt.gz\n",
       "../code/results/qtl/noisy/GTEx/Liver/separateNoise/cis_100000/perm/chr22.addQval.txt.gz"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sgene_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "sgene_df <- map_dfr(sgene_files, fread)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "sgene_df <- sgene_df[q < .05 & str_detect(phenotype_id, \"UP|PR\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>3570</li><li>21</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 3570\n",
       "\\item 21\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 3570\n",
       "2. 21\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 3570   21"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 6 × 21</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>phenotype_id</th><th scope=col>phenotype_chr</th><th scope=col>phenotype_start</th><th scope=col>phenotype_end</th><th scope=col>phenotype_strand</th><th scope=col>num_variants</th><th scope=col>best_nom_dist</th><th scope=col>best_genotype_id</th><th scope=col>best_genotype_chr</th><th scope=col>best_genotype_start</th><th scope=col>⋯</th><th scope=col>dof_true</th><th scope=col>dof_est</th><th scope=col>beta_ml1</th><th scope=col>beta_ml2</th><th scope=col>pval_nom</th><th scope=col>pval_r2</th><th scope=col>slope</th><th scope=col>pval_emp</th><th scope=col>pval_adj</th><th scope=col>q</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>⋯</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1:1082987:1084352:clu_1808_-:PR</td><td>chr1</td><td>1082988</td><td>1084352</td><td>-</td><td>596</td><td>  3532</td><td>chr1_1079456_A_G_b38</td><td>chr1</td><td>1079456</td><td>⋯</td><td>206</td><td>169.900</td><td>1.029230</td><td>52.6246</td><td>4.72618e-07</td><td>0.1161460</td><td>-0.362335</td><td>0.000999001</td><td>1.95822e-04</td><td>0.00517610</td></tr>\n",
       "\t<tr><td>chr1:1090428:1091471:clu_1810_-:PR</td><td>chr1</td><td>1090429</td><td>1091471</td><td>-</td><td>604</td><td>     0</td><td>chr1_1091327_C_A_b38</td><td>chr1</td><td>1091327</td><td>⋯</td><td>206</td><td>170.473</td><td>1.063910</td><td>60.7541</td><td>4.72455e-08</td><td>0.1350960</td><td>-0.444126</td><td>0.000999001</td><td>2.10856e-05</td><td>0.00068731</td></tr>\n",
       "\t<tr><td>chr1:1217688:1218457:clu_1814_-:PR</td><td>chr1</td><td>1217689</td><td>1218457</td><td>-</td><td>594</td><td>     0</td><td>chr1_1217733_G_A_b38</td><td>chr1</td><td>1217733</td><td>⋯</td><td>206</td><td>180.823</td><td>0.985867</td><td>64.7615</td><td>9.74565e-07</td><td>0.1101270</td><td> 0.604832</td><td>0.000999001</td><td>3.28634e-04</td><td>0.00826580</td></tr>\n",
       "\t<tr><td>chr1:1217804:1218457:clu_1814_-:PR</td><td>chr1</td><td>1217805</td><td>1218457</td><td>-</td><td>594</td><td>    72</td><td>chr1_1217733_G_A_b38</td><td>chr1</td><td>1217733</td><td>⋯</td><td>206</td><td>173.863</td><td>1.058870</td><td>59.4772</td><td>8.34233e-07</td><td>0.1114230</td><td>-0.615276</td><td>0.000999001</td><td>2.18888e-04</td><td>0.00572330</td></tr>\n",
       "\t<tr><td>chr1:1389047:1390229:clu_1835_-:UP</td><td>chr1</td><td>1389048</td><td>1390229</td><td>-</td><td>467</td><td>-63641</td><td>chr1_1453870_A_C_b38</td><td>chr1</td><td>1453870</td><td>⋯</td><td>206</td><td>178.993</td><td>0.968296</td><td>37.6122</td><td>1.38712e-05</td><td>0.0878077</td><td>-0.568726</td><td>0.002997000</td><td>2.37384e-03</td><td>0.04552800</td></tr>\n",
       "\t<tr><td>chr1:1642001:1642346:clu_1843_-:PR</td><td>chr1</td><td>1642002</td><td>1642346</td><td>-</td><td>428</td><td> 32112</td><td>chr1_1609890_C_T_b38</td><td>chr1</td><td>1609890</td><td>⋯</td><td>206</td><td>172.917</td><td>0.953546</td><td>40.0320</td><td>1.90325e-06</td><td>0.1045350</td><td>-0.540764</td><td>0.001998000</td><td>7.42839e-04</td><td>0.01646200</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 21\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " phenotype\\_id & phenotype\\_chr & phenotype\\_start & phenotype\\_end & phenotype\\_strand & num\\_variants & best\\_nom\\_dist & best\\_genotype\\_id & best\\_genotype\\_chr & best\\_genotype\\_start & ⋯ & dof\\_true & dof\\_est & beta\\_ml1 & beta\\_ml2 & pval\\_nom & pval\\_r2 & slope & pval\\_emp & pval\\_adj & q\\\\\n",
       " <chr> & <chr> & <int> & <int> & <chr> & <int> & <int> & <chr> & <chr> & <int> & ⋯ & <int> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1:1082987:1084352:clu\\_1808\\_-:PR & chr1 & 1082988 & 1084352 & - & 596 &   3532 & chr1\\_1079456\\_A\\_G\\_b38 & chr1 & 1079456 & ⋯ & 206 & 169.900 & 1.029230 & 52.6246 & 4.72618e-07 & 0.1161460 & -0.362335 & 0.000999001 & 1.95822e-04 & 0.00517610\\\\\n",
       "\t chr1:1090428:1091471:clu\\_1810\\_-:PR & chr1 & 1090429 & 1091471 & - & 604 &      0 & chr1\\_1091327\\_C\\_A\\_b38 & chr1 & 1091327 & ⋯ & 206 & 170.473 & 1.063910 & 60.7541 & 4.72455e-08 & 0.1350960 & -0.444126 & 0.000999001 & 2.10856e-05 & 0.00068731\\\\\n",
       "\t chr1:1217688:1218457:clu\\_1814\\_-:PR & chr1 & 1217689 & 1218457 & - & 594 &      0 & chr1\\_1217733\\_G\\_A\\_b38 & chr1 & 1217733 & ⋯ & 206 & 180.823 & 0.985867 & 64.7615 & 9.74565e-07 & 0.1101270 &  0.604832 & 0.000999001 & 3.28634e-04 & 0.00826580\\\\\n",
       "\t chr1:1217804:1218457:clu\\_1814\\_-:PR & chr1 & 1217805 & 1218457 & - & 594 &     72 & chr1\\_1217733\\_G\\_A\\_b38 & chr1 & 1217733 & ⋯ & 206 & 173.863 & 1.058870 & 59.4772 & 8.34233e-07 & 0.1114230 & -0.615276 & 0.000999001 & 2.18888e-04 & 0.00572330\\\\\n",
       "\t chr1:1389047:1390229:clu\\_1835\\_-:UP & chr1 & 1389048 & 1390229 & - & 467 & -63641 & chr1\\_1453870\\_A\\_C\\_b38 & chr1 & 1453870 & ⋯ & 206 & 178.993 & 0.968296 & 37.6122 & 1.38712e-05 & 0.0878077 & -0.568726 & 0.002997000 & 2.37384e-03 & 0.04552800\\\\\n",
       "\t chr1:1642001:1642346:clu\\_1843\\_-:PR & chr1 & 1642002 & 1642346 & - & 428 &  32112 & chr1\\_1609890\\_C\\_T\\_b38 & chr1 & 1609890 & ⋯ & 206 & 172.917 & 0.953546 & 40.0320 & 1.90325e-06 & 0.1045350 & -0.540764 & 0.001998000 & 7.42839e-04 & 0.01646200\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 21\n",
       "\n",
       "| phenotype_id &lt;chr&gt; | phenotype_chr &lt;chr&gt; | phenotype_start &lt;int&gt; | phenotype_end &lt;int&gt; | phenotype_strand &lt;chr&gt; | num_variants &lt;int&gt; | best_nom_dist &lt;int&gt; | best_genotype_id &lt;chr&gt; | best_genotype_chr &lt;chr&gt; | best_genotype_start &lt;int&gt; | ⋯ ⋯ | dof_true &lt;int&gt; | dof_est &lt;dbl&gt; | beta_ml1 &lt;dbl&gt; | beta_ml2 &lt;dbl&gt; | pval_nom &lt;dbl&gt; | pval_r2 &lt;dbl&gt; | slope &lt;dbl&gt; | pval_emp &lt;dbl&gt; | pval_adj &lt;dbl&gt; | q &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| chr1:1082987:1084352:clu_1808_-:PR | chr1 | 1082988 | 1084352 | - | 596 |   3532 | chr1_1079456_A_G_b38 | chr1 | 1079456 | ⋯ | 206 | 169.900 | 1.029230 | 52.6246 | 4.72618e-07 | 0.1161460 | -0.362335 | 0.000999001 | 1.95822e-04 | 0.00517610 |\n",
       "| chr1:1090428:1091471:clu_1810_-:PR | chr1 | 1090429 | 1091471 | - | 604 |      0 | chr1_1091327_C_A_b38 | chr1 | 1091327 | ⋯ | 206 | 170.473 | 1.063910 | 60.7541 | 4.72455e-08 | 0.1350960 | -0.444126 | 0.000999001 | 2.10856e-05 | 0.00068731 |\n",
       "| chr1:1217688:1218457:clu_1814_-:PR | chr1 | 1217689 | 1218457 | - | 594 |      0 | chr1_1217733_G_A_b38 | chr1 | 1217733 | ⋯ | 206 | 180.823 | 0.985867 | 64.7615 | 9.74565e-07 | 0.1101270 |  0.604832 | 0.000999001 | 3.28634e-04 | 0.00826580 |\n",
       "| chr1:1217804:1218457:clu_1814_-:PR | chr1 | 1217805 | 1218457 | - | 594 |     72 | chr1_1217733_G_A_b38 | chr1 | 1217733 | ⋯ | 206 | 173.863 | 1.058870 | 59.4772 | 8.34233e-07 | 0.1114230 | -0.615276 | 0.000999001 | 2.18888e-04 | 0.00572330 |\n",
       "| chr1:1389047:1390229:clu_1835_-:UP | chr1 | 1389048 | 1390229 | - | 467 | -63641 | chr1_1453870_A_C_b38 | chr1 | 1453870 | ⋯ | 206 | 178.993 | 0.968296 | 37.6122 | 1.38712e-05 | 0.0878077 | -0.568726 | 0.002997000 | 2.37384e-03 | 0.04552800 |\n",
       "| chr1:1642001:1642346:clu_1843_-:PR | chr1 | 1642002 | 1642346 | - | 428 |  32112 | chr1_1609890_C_T_b38 | chr1 | 1609890 | ⋯ | 206 | 172.917 | 0.953546 | 40.0320 | 1.90325e-06 | 0.1045350 | -0.540764 | 0.001998000 | 7.42839e-04 | 0.01646200 |\n",
       "\n"
      ],
      "text/plain": [
       "  phenotype_id                       phenotype_chr phenotype_start\n",
       "1 chr1:1082987:1084352:clu_1808_-:PR chr1          1082988        \n",
       "2 chr1:1090428:1091471:clu_1810_-:PR chr1          1090429        \n",
       "3 chr1:1217688:1218457:clu_1814_-:PR chr1          1217689        \n",
       "4 chr1:1217804:1218457:clu_1814_-:PR chr1          1217805        \n",
       "5 chr1:1389047:1390229:clu_1835_-:UP chr1          1389048        \n",
       "6 chr1:1642001:1642346:clu_1843_-:PR chr1          1642002        \n",
       "  phenotype_end phenotype_strand num_variants best_nom_dist\n",
       "1 1084352       -                596            3532       \n",
       "2 1091471       -                604               0       \n",
       "3 1218457       -                594               0       \n",
       "4 1218457       -                594              72       \n",
       "5 1390229       -                467          -63641       \n",
       "6 1642346       -                428           32112       \n",
       "  best_genotype_id     best_genotype_chr best_genotype_start ⋯ dof_true dof_est\n",
       "1 chr1_1079456_A_G_b38 chr1              1079456             ⋯ 206      169.900\n",
       "2 chr1_1091327_C_A_b38 chr1              1091327             ⋯ 206      170.473\n",
       "3 chr1_1217733_G_A_b38 chr1              1217733             ⋯ 206      180.823\n",
       "4 chr1_1217733_G_A_b38 chr1              1217733             ⋯ 206      173.863\n",
       "5 chr1_1453870_A_C_b38 chr1              1453870             ⋯ 206      178.993\n",
       "6 chr1_1609890_C_T_b38 chr1              1609890             ⋯ 206      172.917\n",
       "  beta_ml1 beta_ml2 pval_nom    pval_r2   slope     pval_emp    pval_adj   \n",
       "1 1.029230 52.6246  4.72618e-07 0.1161460 -0.362335 0.000999001 1.95822e-04\n",
       "2 1.063910 60.7541  4.72455e-08 0.1350960 -0.444126 0.000999001 2.10856e-05\n",
       "3 0.985867 64.7615  9.74565e-07 0.1101270  0.604832 0.000999001 3.28634e-04\n",
       "4 1.058870 59.4772  8.34233e-07 0.1114230 -0.615276 0.000999001 2.18888e-04\n",
       "5 0.968296 37.6122  1.38712e-05 0.0878077 -0.568726 0.002997000 2.37384e-03\n",
       "6 0.953546 40.0320  1.90325e-06 0.1045350 -0.540764 0.001998000 7.42839e-04\n",
       "  q         \n",
       "1 0.00517610\n",
       "2 0.00068731\n",
       "3 0.00826580\n",
       "4 0.00572330\n",
       "5 0.04552800\n",
       "6 0.01646200"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sgene_df %>% dim()\n",
    "sgene_df %>% head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (sos)",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
